{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 - Retreiving and Preparing Text for Machines\n",
    "\n",
    "This week, we begin by \"begging, borrowing and stealing\" text from several\n",
    "contexts of human communication (e.g., PDFs, HTML, Word) and preparing it for\n",
    "machines to \"read\" and analyze. This notebook outlines scraping text from the\n",
    "web, PDF and Word documents. Then we detail \"spidering\" or walking\n",
    "through hyperlinks to build samples of online content, and using APIs,\n",
    "Application Programming Interfaces, provided by webservices to access their\n",
    "content. Along the way, we will use regular expressions, outlined in the\n",
    "reading, to remove unwanted formatting and ornamentation. Finally, we discuss\n",
    "various text encodings, filtering and data structures in which text can be\n",
    "placed for analysis.\n",
    "\n",
    "For this notebook we will be using the following packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Special module written for this class\n",
    "#This provides access to data and to helper functions from previous weeks\n",
    "#import lucem_illud_2020 \n",
    "\n",
    "#All these packages need to be installed from pip\n",
    "import requests #for http requests\n",
    "import bs4 #called `beautifulsoup4`, an html parser\n",
    "import pandas as pd #gives us DataFrames\n",
    "import docx #reading MS doc files, install as `python-docx`\n",
    "import csv\n",
    "\n",
    "#Stuff for pdfs\n",
    "#Install as `pdfminer2`\n",
    "import pdfminer.pdfinterp\n",
    "import pdfminer.converter\n",
    "import pdfminer.layout\n",
    "import pdfminer.pdfpage\n",
    "\n",
    "#These come with Python\n",
    "import re #for regexs\n",
    "import urllib.parse #For joining urls\n",
    "import io #for making http requests look like files\n",
    "import json #For Tumblr API responses\n",
    "import os.path #For checking if files exist\n",
    "import os #For making directories\n",
    "\n",
    "# nltk dependencies \n",
    "import nltk\n",
    "import ssl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will also be working on the following files/urls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "wikipedia_base_url = 'https://en.wikipedia.org'\n",
    "wikipedia_content_analysis = 'https://en.wikipedia.org/wiki/Content_analysis'\n",
    "content_analysis_save = 'wikipedia_content_analysis.html'\n",
    "example_text_file = 'sometextfile.txt'\n",
    "information_extraction_pdf = 'https://github.com/Computational-Content-Analysis-2018/Data-Files/raw/master/1-intro/Content%20Analysis%2018.pdf'\n",
    "example_docx = 'https://github.com/Computational-Content-Analysis-2018/Data-Files/raw/master/1-intro/macs6000_connecting_to_midway.docx'\n",
    "example_docx_save = 'example.docx'  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scraping\n",
    "\n",
    "Before we can start analyzing content we need to obtain it. Sometimes it will be\n",
    "provided to us from a pre-curated text archive, but sometimes we will need to\n",
    "download it. As a starting example we will attempt to download the wikipedia\n",
    "page on content analysis. The page is located at [https://en.wikipedia.org/wiki/\n",
    "Content_analysis](https://en.wikipedia.org/wiki/Content_analysis) so lets start\n",
    "with that.\n",
    "\n",
    "We can do this by making an HTTP GET request to that url, a GET request is\n",
    "simply a request to the server to provide the contents given by some url. The\n",
    "other request we will be using in this class is called a POST request and\n",
    "requests the server to take some content we provide. While the Python standard\n",
    "library does have the ability do make GET requests we will be using the\n",
    "[_requests_](http://docs.python-requests.org/en/master/) package as it is _'the\n",
    "only Non-GMO HTTP library for Python'_...also it provides a nicer interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Response [200]>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#wikipedia_content_analysis = 'https://en.wikipedia.org/wiki/Content_analysis'\n",
    "requests.get(wikipedia_content_analysis)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`'Response [200]'` means the server responded with what we asked for. If you get\n",
    "another number (e.g. 404) it likely means there was some kind of error, these\n",
    "codes are called HTTP response codes and a list of them can be found\n",
    "[here](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes). The response\n",
    "object contains all the data the server sent including the website's contents\n",
    "and the HTTP header. We are interested in the contents which we can access with\n",
    "the `.text` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html>\n",
      "<html class=\"client-nojs\" lang=\"en\" dir=\"ltr\">\n",
      "<head>\n",
      "<meta charset=\"UTF-8\"/>\n",
      "<title>Content analysis - Wikipedia</title>\n",
      "<script>document.documentElement.className=\"client-js\";RLCONF={\"wgBreakFrames\":!1,\"wgSeparatorTransformTable\":[\"\",\"\"],\"wgDigitTransformTable\":[\"\",\"\"],\"wgDefaultDateFormat\":\"dmy\",\"wgMonthNames\":[\"\",\"January\",\"February\",\"March\",\"April\",\"May\",\"June\",\"July\",\"August\",\"September\",\"October\",\"November\",\"December\"],\"wgMonthNamesShort\":[\"\",\"Jan\",\"Feb\",\"Mar\",\"Apr\",\"May\",\"Jun\",\"Jul\",\"Aug\",\"Sep\",\"Oct\",\"Nov\",\"Dec\"],\"wgRequestId\":\"XitMrApAAEIAAF5EbAQAAAAF\",\"wgCSPNonce\":!1,\"wgCanonicalNamespace\":\"\",\"wgCanonicalSpecialPageName\":!1,\"wgNamespaceNumber\":0,\"wgPageName\":\"Content_analysis\",\"wgTitle\":\"Content analysis\",\"wgCurRevisionId\":937401412,\"wgRevisionId\":937401412,\"wgArticleId\":473317,\"wgIsArticle\":!0,\"wgIsRedirect\":!1,\"wgAction\":\"view\",\"wgUserName\":null,\"wgUserGroups\":[\"*\"],\"wgCategories\":[\"Articles needing expert attention with no reason or talk parameter\",\"Article\n"
     ]
    }
   ],
   "source": [
    "wikiContentRequest = requests.get(wikipedia_content_analysis)\n",
    "print(wikiContentRequest.text[:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is not what we were looking for, because it is the start of the HTML that\n",
    "makes up the website. This is HTML and is meant to be read by computers. Luckily\n",
    "we have a computer to parse it for us. To do the parsing we will use [_Beautiful\n",
    "Soup_](https://www.crummy.com/software/BeautifulSoup/) which is a better parser\n",
    "than the one in the standard library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "Content analysis - Wikipedia\n",
      "document.documentElement.className=\"client-js\";RLCONF={\"wgBreakFrames\":!1,\"wgSeparatorTransformTable\":[\"\",\"\"],\"wgDigitTransformTable\":[\"\",\"\"],\"wgDefaultDateFormat\":\"dm\n"
     ]
    }
   ],
   "source": [
    "wikiContentSoup = bs4.BeautifulSoup(wikiContentRequest.text, 'html.parser')\n",
    "print(wikiContentSoup.text[:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is better but there's still random whitespace and we have more than just\n",
    "the text of the article. This is because what we requested is the whole webpage,\n",
    "not just the text for the article.\n",
    "\n",
    "We want to extract only the text we care about, and in order to do this we will\n",
    "need to inspect the html. One way to do this is simply to go to the website with\n",
    "a browser and use its inspection or view source tool. If javascript or other\n",
    "dynamic loading occurs on the page, however, it is likely that what Python\n",
    "receives is not what you will see, so we will need to inspect what Python\n",
    "receives. To do this we can save the html `requests` obtained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#content_analysis_save = 'wikipedia_content_analysis.html'\n",
    "\n",
    "with open(content_analysis_save, mode='w', encoding='utf-8') as f:\n",
    "    f.write(wikiContentRequest.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets open the file (`wikipedia_content_analysis.html`) we just created with\n",
    "a web browser. It should look sort of like the original but without the images\n",
    "and formatting.\n",
    "\n",
    "As there is very little standardization on structuring webpages, figuring out\n",
    "how best to extract what you want is an art. Looking at this page it looks like\n",
    "all the main textual content is inside `<p>`(paragraph) tags within the `<body>`\n",
    "tag."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Content analysis is a research method for studying documents and communication artifacts, which might be texts of various formats, pictures, audio or video. Social scientists use content analysis to examine patterns in communication in a replicable and systematic manner.[1] One of the key advantages of using content analysis to analyse social phenomena is its non-invasive nature, in contrast to simulating social experiences or collecting survey answers.\n",
      "\n",
      "Practices and philosophies of content analysis vary between academic disciplines. They all involve systematic reading or observation of texts or artifacts which are assigned labels (sometimes called codes) to indicate the presence of interesting, meaningful pieces of content.[2][3] By systematically labeling the content of a set of texts, researchers can analyse patterns of content quantitatively using statistical methods, or use qualitative methods to analyse meanings of content within texts.\n",
      "\n",
      "Computers are increasingly used in content analysis to automate the labeling (or coding) of documents. Simple computational techniques can provide descriptive data such as word frequencies and document lengths. Machine learning classifiers can greatly increase the number of texts that can be labeled, but the scientific utility of doing so is a matter of debate. Further, numerous computer-aided text analysis (CATA) computer programs are available that analyze text for pre-determined linguistic, semantic, and psychological characteristics.[4]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "contentPTags = wikiContentSoup.body.findAll('p')\n",
    "for pTag in contentPTags[:3]:\n",
    "    print(pTag.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have all the text from the page, split up by paragraph. If we wanted to\n",
    "get the section headers or references as well it would require a bit more work,\n",
    "but is doable.\n",
    "\n",
    "There is one more thing we might want to do before sending this text to be\n",
    "processed, remove the references indicators (`[2]`, `[3]` , etc). To do this we\n",
    "can use a short regular expression (regex)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                       paragraph-text\n",
      "0   Content analysis is a research method for stud...\n",
      "1   Practices and philosophies of content analysis...\n",
      "2   Computers are increasingly used in content ana...\n",
      "3   Content analysis is best understood as a broad...\n",
      "4   The simplest and most objective form of conten...\n",
      "5   A further step in analysis is the distinction ...\n",
      "6   More generally, content analysis is research u...\n",
      "7   By having contents of communication available ...\n",
      "8   Robert Weber notes: \"To make valid inferences ...\n",
      "9   There are five types of texts in content analy...\n",
      "10  Over the years, content analysis has been appl...\n",
      "11  In recent times, particularly with the advent ...\n",
      "12  Quantitative content analysis has enjoyed a re...\n",
      "13  Content analysis can also be described as stud...\n",
      "14  Holsti groups fifteen uses of content analysis...\n",
      "15  He also places these uses into the context of ...\n",
      "16  The following table shows fifteen uses of cont...\n",
      "17  The process of the initial coding scheme or ap...\n",
      "18  With either approach above, immersing oneself ...\n",
      "19                                                 \\n\n"
     ]
    }
   ],
   "source": [
    "contentParagraphs = []\n",
    "for pTag in contentPTags:\n",
    "    #strings starting with r are raw so their \\'s are not modifier characters\n",
    "    #If we didn't start with r the string would be: '\\\\[\\\\d+\\\\]'\n",
    "    contentParagraphs.append(re.sub(r'\\[\\d+\\]', '', pTag.text))\n",
    "\n",
    "#convert to a DataFrame\n",
    "contentParagraphsDF = pd.DataFrame({'paragraph-text' : contentParagraphs})\n",
    "print(contentParagraphsDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a `DataFrame` containing all relevant text from the page ready to be\n",
    "processed\n",
    "\n",
    "If you are not familiar with regex, it is a way of specifying searches in text.\n",
    "A regex engine takes in the search pattern, in the above case `'\\[\\d+\\]'` and\n",
    "some string, the paragraph texts. Then it reads the input string one character\n",
    "at a time checking if it matches the search. Here the regex `'\\d'` matches\n",
    "number characters (while `'\\['` and `'\\]'` capture the braces on either side)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<re.Match object; span=(36, 37), match='2'>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "findNumber = r'\\d'\n",
    "regexResults = re.search(findNumber, 'not a number, not a number, numbers 2134567890, not a number')\n",
    "regexResults"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Python the regex package (`re`) usually returns `Match` objects (you can have\n",
    "multiple pattern hits in a a single `Match`), to get the string that matched our\n",
    "pattern we can use the `.group()` method, and as we want the first one we will\n",
    "ask for the 0'th group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "print(regexResults.group(0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That gives us the first number, if we wanted the whole block of numbers we can\n",
    "add a wildcard `'+'` which requests 1 or more instances of the preceding\n",
    "character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2134567890\n"
     ]
    }
   ],
   "source": [
    "findNumbers = r'\\d+'\n",
    "regexResults = re.search(findNumbers, 'not a number, not a number, numbers 2134567890, not a number')\n",
    "print(regexResults.group(0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have the whole block of numbers, there are a huge number of special\n",
    "characters in regex, for the full description of Python's implementation look at\n",
    "the [re docs](https://docs.python.org/3/library/re.html) there is also a short\n",
    "[tutorial](https://docs.python.org/3/howto/regex.html#regex-howto)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <span style=\"color:red\">Section 1</span>\n",
    "<span style=\"color:red\">Construct cells immediately below this that describe and download webcontent relating to your anticipated final project. Use beautiful soup and at least five regular expressions to extract relevant, nontrivial *chunks* of that content (e.g., cleaned sentences, paragraphs, etc.) to a pandas `Dataframe`.</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So I want to scrape two things: one, the title and abstract of every paper who's PMID is in `pmids_df` and two, the full text of every article mentioned in `pmids_df`. For the purpose of this assignment, I'll just try and assemble the corpra of tites and abstracts. (I'll do full text later.) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pmids_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-d6e06c0d1d46>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpmids_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'pmids_df' is not defined"
     ]
    }
   ],
   "source": [
    "pmids_df[0].iloc[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_pmid_url = 'https://www.ncbi.nlm.nih.gov/pubmed/?term=30784590'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "requests.get(first_pmid_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "firstPMIDContentRequest = requests.get(first_pmid_url)\n",
    "print(firstPMIDContentRequest.text[:1000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "firstPMIDContentSoup = bs4.BeautifulSoup(firstPMIDContentRequest.text, 'html.parser')\n",
    "print(firstPMIDContentSoup.text[:127])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The Abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "abstract = firstPMIDContentSoup.find(\"div\", class_='abstr')\n",
    "print(abstract.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5 Regular Expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_num = re.sub(\" \\d+\", \" \", abstract.text)\n",
    "print(no_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5 regex -- 1.abstract 2. copyright symbol 3.year 3. basically everything after copyright word \n",
    "drop_abs = re.sub(\"Abstract\", \"\", no_num)\n",
    "drop_abs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_copy = re.sub(\"Copyright\", \"\", drop_abs)\n",
    "drop_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_symb = re.sub(\"\\u00a9\", \"\", drop_copy)\n",
    "drop_symb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean = re.sub(\"The Authors. Published by Elsevier Inc. All rights reserved.\", \"\", drop_symb)\n",
    "clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full = []\n",
    "full.append(clean)\n",
    "abstract_df = pd.DataFrame({'abstract': full })\n",
    "abstract_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The Title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = firstPMIDContentSoup.findAll(\"h1\")\n",
    "PMIDtitle = [title[1].text] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title_df = pd.DataFrame({'title': PMIDtitle})\n",
    "title_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Spidering\n",
    "\n",
    "What if we want to to get a bunch of different pages from wikipedia. We would\n",
    "need to get the url for each of the pages we want. Typically, we want pages that\n",
    "are linked to by other pages and so we will need to parse pages and identify the\n",
    "links. Right now we will be retrieving all links in the body of the content\n",
    "analysis page.\n",
    "\n",
    "To do this we will need to find all the `<a>` (anchor) tags with `href`s\n",
    "(hyperlink references) inside of `<p>` tags. `href` can have many\n",
    "[different](http://stackoverflow.com/questions/4855168/what-is-href-and-why-is-\n",
    "it-used) [forms](https://en.wikipedia.org/wiki/Hyperlink#Hyperlinks_in_HTML) so\n",
    "dealing with them can be tricky, but generally, you will want to extract\n",
    "absolute or relative links. An absolute link is one you can follow without\n",
    "modification, while a relative link requires a base url that you will then\n",
    "append. Wikipedia uses relative urls for its internal links: below is an example\n",
    "for dealing with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#wikipedia_base_url = 'https://en.wikipedia.org'\n",
    "\n",
    "otherPAgeURLS = []\n",
    "#We also want to know where the links come from so we also will get:\n",
    "#the paragraph number\n",
    "#the word the link is in\n",
    "for paragraphNum, pTag in enumerate(contentPTags):\n",
    "    #we only want hrefs that link to wiki pages\n",
    "    tagLinks = pTag.findAll('a', href=re.compile('/wiki/'), class_=False)\n",
    "    for aTag in tagLinks:\n",
    "        #We need to extract the url from the <a> tag\n",
    "        relurl = aTag.get('href')\n",
    "        linkText = aTag.text\n",
    "        #wikipedia_base_url is the base we can use the urllib joining function to merge them\n",
    "        #Giving a nice structured tupe like this means we can use tuple expansion later\n",
    "        otherPAgeURLS.append((\n",
    "            urllib.parse.urljoin(wikipedia_base_url, relurl),\n",
    "            paragraphNum,\n",
    "            linkText,\n",
    "        ))\n",
    "print(otherPAgeURLS[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will be adding these new texts to our DataFrame `contentParagraphsDF` so we\n",
    "will need to add 2 more columns to keep track of paragraph numbers and sources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contentParagraphsDF['source'] = [wikipedia_content_analysis] * len(contentParagraphsDF['paragraph-text'])\n",
    "contentParagraphsDF['paragraph-number'] = range(len(contentParagraphsDF['paragraph-text']))\n",
    "\n",
    "contentParagraphsDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we can add two more columns to our `Dataframe` and define a function to\n",
    "parse\n",
    "each linked page and add its text to our DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contentParagraphsDF['source-paragraph-number'] = [None] * len(contentParagraphsDF['paragraph-text'])\n",
    "contentParagraphsDF['source-paragraph-text'] = [None] * len(contentParagraphsDF['paragraph-text'])\n",
    "\n",
    "def getTextFromWikiPage(targetURL, sourceParNum, sourceText):\n",
    "    #Make a dict to store data before adding it to the DataFrame\n",
    "    parsDict = {'source' : [], 'paragraph-number' : [], 'paragraph-text' : [], 'source-paragraph-number' : [],  'source-paragraph-text' : []}\n",
    "    #Now we get the page\n",
    "    r = requests.get(targetURL)\n",
    "    soup = bs4.BeautifulSoup(r.text, 'html.parser')\n",
    "    #enumerating gives use the paragraph number\n",
    "    for parNum, pTag in enumerate(soup.body.findAll('p')):\n",
    "        #same regex as before\n",
    "        parsDict['paragraph-text'].append(re.sub(r'\\[\\d+\\]', '', pTag.text))\n",
    "        parsDict['paragraph-number'].append(parNum)\n",
    "        parsDict['source'].append(targetURL)\n",
    "        parsDict['source-paragraph-number'].append(sourceParNum)\n",
    "        parsDict['source-paragraph-text'].append(sourceText)\n",
    "    return pandas.DataFrame(parsDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And run it on our list of link tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for urlTuple in otherPAgeURLS[:3]:\n",
    "    #ignore_index means the indices will not be reset after each append\n",
    "    contentParagraphsDF = contentParagraphsDF.append(getTextFromWikiPage(*urlTuple),ignore_index=True)\n",
    "contentParagraphsDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# <span style=\"color:red\">Section 2</span>\n",
    "<span style=\"color:red\">Construct cells immediately below this that spider webcontent from another site with content relating to your anticipated final project. Specifically, identify urls on a core page, then follow and extract content from them into a pandas `Dataframe`. In addition, demonstrate a *recursive* spider, which follows more than one level of links (i.e., follows links from a site, then follows links on followed sites to new sites, etc.), making sure to define a reasonable endpoint so that you do not wander the web forever :-).</span>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['http://www.ncbi.nlm.nih.gov/pubmed/30784590',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30692680',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30659290',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30609404',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30445657',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1264230',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1113041',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5425661',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5724688',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5582959',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30864334',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30592451',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30275053',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30240647',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29777175',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10014315',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10011709',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3872519',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6706255',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5089140',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30474576',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30449211',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30345364',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30179987',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29580902',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13534298',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13425430',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12990522',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14907503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14861110',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926926',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926925',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926924',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926922',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30922499',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20996781',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19873382',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18741625',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16744786',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20891193',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30920417',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30917518',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30897527',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30897179',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30890620',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21379958',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21379949',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21379938',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21372959',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21372866',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5457638',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5570771',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5570780',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16592017',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4509649',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29118206',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29190358',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29255778',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29930110',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30545854',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30894395',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30188541',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29610250',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29474918',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29423234',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8756646',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7726173',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7913883',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8136842',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7506603',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30922461',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30917804',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30903564',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30895787',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30892738',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1368750',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2371972',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2560594',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2789738',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2766066',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30854722',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30846484',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30845981',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30626636',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30519676',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5689503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6066721',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5921379',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5838400',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13084255',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30735129',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30225343',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30191425',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29597197',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28247015',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21307056',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18628294',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17991292',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17947320',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9040042',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30724261',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30409642',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30242936',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30055435',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29908446',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8079737',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7511137',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8343277',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1745918',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16664890',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30914511',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30659188',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29959346',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29859085',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29774858',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/935247',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4132765',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4992408',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5414930',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4979525',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30856613',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28689370',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28034351',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26315828',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21030160',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2348958',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6702971',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7105735',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7035460',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1171673',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30872237',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30664273',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30513724',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30224652',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30014848',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10731132',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10731134',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10581034',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9847238',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7906415',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926852',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926491',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30925992',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30925194',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30925118',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13231192',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20283073',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17104818',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21031891',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21031875',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29773677',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29554099',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29325522',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28610618',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28350385',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1619366',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1611868',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1569370',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2052090',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2188686',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30928653',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30891863',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30846474',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30830396',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30691708',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6850144',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7311872',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7248561',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7248560',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7214054',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28275956',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27694391',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26505614',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15251069',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10835600',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6528998',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6466773',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6742429',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13805995',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13449371',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28609229',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27575953',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26042762',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24667040',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24040834',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/225101',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/357434',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/340459',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/340458',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/340457',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30778087',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30699870',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30603797',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30531653',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30431293',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6193062',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6168574',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/232487',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/571412',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/730277',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30879167',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30862659',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30852185',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30842308',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30832630',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20242920',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19991283',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17856943',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16577536',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21379478',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30923040',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30837567',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30478097',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30478095',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30429608',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2506560',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3365248',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3132203',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3674400',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3779697',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30339633',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28067713',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25710421',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24919177',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24080131',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4060567',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7379123',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/412833',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/791372',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/820684',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30927043',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926791',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30925871',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30915493',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30910026',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18148946',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18127701',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15409300',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21065619',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21023268',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30778196',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27918567',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27244377',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27082371',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27006089',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/926954',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/408650',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/979484',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4217059',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13586729',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23404370',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22521665',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17974005',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17696195',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16940977',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9126902',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7553651',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7717450',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7767486',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8137236',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21455490',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21293372',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21270892',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20966253',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20489017',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11237014',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11237011',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11157790',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10738668',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10391214',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28168274',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26283309',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27424191',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27129864',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27035062',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11130714',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10962990',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10784281',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10655503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8889549',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30311266',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30026804',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29936130',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29856123',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28864551',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6893989',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6893051',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5157368',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4162039',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4221922',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30889022',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30878373',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30770879',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30629992',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29895339',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24170010',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7766333',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8119749',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6299660',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12336179',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30902869',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30781698',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30568658',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30415127',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30368040',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1312140',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3237247',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3504447',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3857806',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6811981',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30096302',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30053901',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29625053',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29625052',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29625049',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11494768',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11237011',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11157790',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9750196',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9521921',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30902578',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30857560',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30842212',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30632250',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30598455',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2265608',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2575066',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3678820',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6392851',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6297748',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30792401',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30791106',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30320429',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29657180',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28946846',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7837392',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7837387',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7894491',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7827164',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1693529',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30788456',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30304494',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30266789',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30177229',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30154102',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8609628',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2726796',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2960378',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3448608',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7326392',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30877103',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30833417',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30744412',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30705421',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30617194',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10668963',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10668960',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8849440',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9063638',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14425247',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924540',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924539',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30918891',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30913092',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30871224',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19971914',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19971896',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19971879',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16992870',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17856195',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23594743',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21490962',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19325880',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18477386',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17943131',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6350972',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6302707',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6260403',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6247112',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/755388',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30921102',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30912126',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30911048',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30905472',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30889373',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17231783',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17231780',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17231777',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17231675',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20750607',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30921272',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30902551',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30890493',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30890492',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30890491',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14101961',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13959157',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14015502',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14008439',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13061454',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15921143',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8114212',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8239852',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1951806',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1853861',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3190261',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2957397',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2956294',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4078095',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6367669',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27044459',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24687478',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22753198',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16509285',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15198273',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2053908',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2396332',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3212846',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4687834',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4661297',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30728404',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30688400',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28912378',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28827708',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28461402',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11729165',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11361330',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11277441',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19281376',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8552036',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30701077',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30607981',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30253119',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30071306',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29881121',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12468126',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12387852',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12207002',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11934736',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11788634',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28818445',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21283829',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20600807',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20416506',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20380691',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6203096',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6306587',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7133994',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1213674',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14404720',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30928021',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924967',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924297',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924134',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30919880',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20756862',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20754607',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20753813',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20751482',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30434084',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30920222',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30916311',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30885009',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30882682',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30868805',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14207227',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13849378',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13345802',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13328118',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16654441',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30481831',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30411598',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29088705',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28934964',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28384336',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2398009',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2354048',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2340149',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3221254',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1156720',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30867456',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30475855',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30265702',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29945178',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29850621',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12547909',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12142439',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11948594',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11125101',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10568754',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29636096',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29504913',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28778154',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26393794',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26161537',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8808573',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7922689',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5509291',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14236963',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14003908',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30909870',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30880156',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30880041',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30809187',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30806469',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14048412',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13956846',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13956845',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13429770',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13136171',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30836076',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30700871',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30698640',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30668635',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30647110',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14784824',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18124447',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18885184',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17647974',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30330878',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30872971',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30565092',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30558353',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30517796',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30517793',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14941824',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14930176',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14772731',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19970473',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19990145',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30785314',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30307741',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30010385',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29896978',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29705456',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5570368',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5449614',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14442915',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13360300',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21407503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30845982',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30637004',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1339803',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30928954',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926777',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18134348',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18117962',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18938432',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18902412',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18902411',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30213880',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29624377',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15489334',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15489320',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12477932',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4859177',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4776997',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4542931',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4704702',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5132403',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30875938',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30439956',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30417113',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30379134',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30355749',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16161146',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16045478',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15943967',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15659098',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14121659',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30698752',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30581286',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28138850',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25424553',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22102577',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2564664',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2563586',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2902571',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2901068',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2895913',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926139',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30907453',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30844082',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30246120',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30243819',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2099718',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3202581',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2958840',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3600485',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3477719',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29453291',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27047762',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24817670',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24123261',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23824657',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2247076',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2162476',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2170811',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2909529',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3221863',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30784590',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30669388',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30455444',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30275530',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30226838',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13818817',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13493338',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13463272',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13295553',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14372983',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30917344',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30912775',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30905312',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30902503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30888153',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6224662',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6224653',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7338149',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6455267',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7227122',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23534382',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22995636',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22620680',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22360692',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22214747',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6466719',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6331853',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6830890',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7074161',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7191331',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30674655',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30298651',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29761043',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28463984',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27022768',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7678659',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2071936',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1999496',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2001227',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2332972',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27072794',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24000942',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19706180',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19654113',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17510325',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14980218',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12149135',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10977085',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10731132',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9521932',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30921535',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30826122',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30809816',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30439455',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30423352',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14209033',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14014830',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13930982',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14465901',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14465900',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30701084',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29246971',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29019350',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28760345',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28466843',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12973360',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12899730',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11432817',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10393112',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10052458',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27646542',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24586521',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23691516',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12147455',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11601685',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5295722',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13173027',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13043099',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13023594',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30495794',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30690522',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30533761',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30403025',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30075815',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29853263',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4445438',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4718013',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4559045',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4555179',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13394881',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30867611',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30849376',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30849323',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30735661',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30712874',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2884728',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3470801',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2876423',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3752091',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2884068',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30927611',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924134',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30923243',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30921127',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30919157',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30163100',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29638605',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20745358',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20745327',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30491266',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30890632',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30875028',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30867156',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30853888',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30823441',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17856542',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17866941',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17866380',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17865859',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17865784',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30851431',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30661756',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30651371',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30635271',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30559433',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13476960',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13321211',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13184683',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13097994',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14917969',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30894680',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30872493',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30833764',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30736069',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30735633',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10043583',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3753762',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4370560',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4370023',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4812038',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30375276',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30136037',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30008075',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29903773',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29624942',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/12500486',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11578293',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9849319',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1582263',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14072027',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18091988',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18072598',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16952911',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16235556',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16204196',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10984453',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9931339',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8976060',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8825759',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8069416',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30933983',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30933671',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30930099',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30924380',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30913208',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17783533',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17819334',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17817247',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17815970',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30494490',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27789023',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23514125',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29758783',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21502420',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20854096',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/3919393',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6591177',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6986389',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/377581',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/22399',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30069064',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28609482',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27899644',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27643465',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27271295',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7659016',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8001908',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8406411',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1357691',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1685753',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30777925',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30737146',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30635745',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30605687',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30257204',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9976484',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10011123',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10053270',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10004256',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10044155',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30659503',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30341961',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30328199',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29803338',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29101024',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6337257',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7050346',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6946226',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/6938498',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/4135212',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29967967',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29471711',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29214968',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28272539',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27649827',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11691856',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11483579',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11015564',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10735275',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14825243',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29727481',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29271913',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28017738',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26445184',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26061457',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7480898',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1924902',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2392579',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/533333',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1145397',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30928302',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30895423',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30820864',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30789005',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30728542',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17808411',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17842591',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17842546',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17732508',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30493732',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30793898',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30352331',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30301924',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29771377',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29100409',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9665479',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9582532',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8972753',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8763264',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/2039195',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926973',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30820047',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30721968',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30698716',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30679814',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9897026',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/19724441',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/1101393',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5523599',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13478613',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11913724',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11840487',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11346797',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/11035790',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10973253',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8896568',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8696331',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8826440',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7633441',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7987295',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30414275',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30220471',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24733792',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24667040',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/21217814',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10631152',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10391209',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9758611',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8693731',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/7824496',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30704421',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30638416',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30629342',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30626376',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30453808',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20781448',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20778163',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20775166',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20772347',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20768152',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26432245',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26014595',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24980144',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24634402',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/23329035',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20396055',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16592864',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14015812',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13938989',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13938988',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30923749',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30674342',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30122434',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29905971',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29855382',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/16057139',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/5393841',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/14315168',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17833878',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/13617309',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30923026',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30918009',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30912961',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30887440',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30865886',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30333324',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30192041',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30177862',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30156454',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30125349',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30907406',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30878477',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30840506',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30808979',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30801747',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30545937',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30502264',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30509563',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30485262',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30401773',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29769526',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28270613',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27191595',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25700176',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25645699',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/18309376',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17804842',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/17804836',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15718098',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15177568',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30930166',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30377334',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30312519',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29895857',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29884854',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27631725',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26416980',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26187847',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25205790',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25066379',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30928167',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30926219',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30916871',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30904800',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30893370',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30806582',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30791507',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30785103',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30773700',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30772070',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28103233',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/27054066',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25072583',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15620887',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/15123810',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10592193',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10508518',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/10330125',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/9294446',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/8449886',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30573480',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30420483',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30409445',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30249784',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30198573',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29053898',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28829531',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28704407',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28673965',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/28636076',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/26324928',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25149520',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/25016412',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24879607',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/24812699',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20927367',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20851965',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20547848',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20482647',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/20331772',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30919672',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30870384',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30802200',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30779925',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/30715965',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29899128',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29896617',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29896617',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29716655',\n",
       " 'http://www.ncbi.nlm.nih.gov/pubmed/29617135',\n",
       " ...]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_links = []\n",
    "for pmid in pmids_list: \n",
    "    all_links.append(\"http://www.ncbi.nlm.nih.gov/pubmed/%s\" % pmid)\n",
    "    \n",
    "all_links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    376\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Python 2.7, use buffering of HTTP responses\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 377\u001b[0;31m                 \u001b[0mhttplib_response\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffering\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    378\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Python 3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: getresponse() got an unexpected keyword argument 'buffering'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-c95a8012796c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mfull\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mlink\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mall_links\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mPMIDContentRequest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrequests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlink\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mPMIDContentSoup\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbs4\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mBeautifulSoup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPMIDContentRequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'html.parser'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mabstract\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPMIDContentSoup\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"div\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclass_\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'abstr'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/api.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m     \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefault\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'allow_redirects'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'get'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/api.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 60\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     61\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    531\u001b[0m         }\n\u001b[1;32m    532\u001b[0m         \u001b[0msend_kwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 533\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    534\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    535\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    666\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;31m# Resolve redirects if allowed.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 668\u001b[0;31m         \u001b[0mhistory\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mresp\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mresp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgen\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mallow_redirects\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    669\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    670\u001b[0m         \u001b[0;31m# Shuffle things around if there's history.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    666\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;31m# Resolve redirects if allowed.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 668\u001b[0;31m         \u001b[0mhistory\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mresp\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mresp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgen\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mallow_redirects\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    669\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    670\u001b[0m         \u001b[0;31m# Shuffle things around if there's history.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mresolve_redirects\u001b[0;34m(self, resp, req, stream, timeout, verify, cert, proxies, yield_requests, **adapter_kwargs)\u001b[0m\n\u001b[1;32m    237\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    238\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 239\u001b[0;31m                 resp = self.send(\n\u001b[0m\u001b[1;32m    240\u001b[0m                     \u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    241\u001b[0m                     \u001b[0mstream\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    644\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    645\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 646\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    648\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    438\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mchunked\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m                 resp = conn.urlopen(\n\u001b[0m\u001b[1;32m    440\u001b[0m                     \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m                     \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    595\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    596\u001b[0m             \u001b[0;31m# Make the request on the httplib connection object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 597\u001b[0;31m             httplib_response = self._make_request(conn, method, url,\n\u001b[0m\u001b[1;32m    598\u001b[0m                                                   \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout_obj\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    599\u001b[0m                                                   \u001b[0mbody\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    378\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Python 3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    379\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 380\u001b[0;31m                     \u001b[0mhttplib_response\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    381\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    382\u001b[0m                     \u001b[0;31m# Remove the TypeError from the exception chain in Python 3;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1321\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1322\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1323\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1324\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;31m# read until we get a non-100 response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 303\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    304\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 264\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    265\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    266\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mLineTooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"status line\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 669\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    670\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1239\u001b[0m                   \u001b[0;34m\"non-zero flags not allowed in calls to recv_into() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1240\u001b[0m                   self.__class__)\n\u001b[0;32m-> 1241\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1242\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1243\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1097\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbuffer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1099\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1100\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1101\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "full = []\n",
    "for link in all_links: \n",
    "    PMIDContentRequest = requests.get(link)\n",
    "    PMIDContentSoup = bs4.BeautifulSoup(PMIDContentRequest.text, 'html.parser')\n",
    "    abstract = PMIDContentSoup.find(\"div\", class_='abstr')\n",
    "    title = PMIDContentSoup.findAll(\"h1\")\n",
    "    #PMIDtitle = [title[1].text] \n",
    "    full.append({'abstract': abstract,\n",
    "                'title': title}) #very unclean wo regex \n",
    "full_df = pd.DataFrame(full)\n",
    "full_df #this might be too much for website"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recursive Spider\n",
    "To demonstrate the recursive spider, I'll use the link I did in the scrape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compiled w additions from the above \n",
    "def recursive_spider(targetURL, sourceParNum, sourceText, limit,i=0):\n",
    "    fullDict = {'source' : [], 'paragraph-number' : [], 'text' : [], 'source-paragraph-number' : [],  'source-paragraph-text' : []}\n",
    "    urls = []\n",
    "\n",
    "    print('getting page {}'.format(i))\n",
    "    req = requests.get(targetURL)\n",
    "    soup = bs4.BeautifulSoup(req.text, 'html.parser')\n",
    "\n",
    "    for pNum, pTag in enumerate(soup.body.findAll('p')):        \n",
    "        #getting paragraph\n",
    "        fullDict['text'].append(pTag.text)\n",
    "        fullDict['paragraph-number'].append(pNum)\n",
    "        fullDict['source'].append(targetURL) \n",
    "        fullDict['source-paragraph-number'].append(sourceParNum)\n",
    "        fullDict['source-paragraph-text'].append(sourceText)\n",
    "        \n",
    "        #getting links\n",
    "        tag_links = pTag.findAll('a', href=re.compile('/ncbi/'), class_=False)\n",
    "        for link in tag_links:\n",
    "            relurl = link.get('href')\n",
    "            link_text = link.text.strip()\n",
    "            urls.append((\n",
    "                urllib.parse.urljoin(wikipedia_base_url, relurl),\n",
    "                pNum,\n",
    "                link_text))\n",
    "    df = pd.DataFrame(fullDict)\n",
    "    \n",
    "    print(urls[:limit])\n",
    "    for urlTuple in urls[:limit]:\n",
    "        if i>limit: \n",
    "            # limit the number of recursive depth\n",
    "            break\n",
    "        else: \n",
    "            i += 1\n",
    "            # so as not to reset index after every append\n",
    "            df = df.append(recursive_spider(*urlTuple,limit,i),ignore_index=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "website = 'https://www.ncbi.nlm.nih.gov/pubmed/?term=30784590'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = recursive_spider(website, None, None, limit = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['source'].unique()\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## API (Tumblr)\n",
    "\n",
    "Generally website owners do not like you scraping their sites. If done badly,\n",
    "scarping can act like a DOS attack so you should be careful how often you make\n",
    "calls to a site. Some sites want automated tools to access their data, so they\n",
    "create [application programming interface\n",
    "(APIs)](https://en.wikipedia.org/wiki/Application_programming_interface). An API\n",
    "specifies a procedure for an application (or script) to access their data. Often\n",
    "this is though a [representational state transfer\n",
    "(REST)](https://en.wikipedia.org/wiki/Representational_state_transfer) web\n",
    "service, which just means if you make correctly formatted HTTP requests they\n",
    "will return nicely formatted data.\n",
    "\n",
    "A nice example for us to study is [Tumblr](https://www.tumblr.com), they have a\n",
    "[simple RESTful API](https://www.tumblr.com/docs/en/api/v1) that allows you to\n",
    "read posts without any complicated html parsing.\n",
    "\n",
    "We can get the first 20 posts from a blog by making an http GET request to\n",
    "`'http://{blog}.tumblr.com/api/read/json'`, were `{blog}` is the name of the\n",
    "target blog. Lets try and get the posts from [http://lolcats-lol-\n",
    "cat.tumblr.com/](http://lolcats-lol-cat.tumblr.com/) (Note the blog says at the\n",
    "top 'One hour one pic lolcats', but the canonical name that Tumblr uses is in\n",
    "the URL 'lolcats-lol-cat')."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tumblrAPItarget = 'http://{}.tumblr.com/api/read/json'\n",
    "\n",
    "r = requests.get(tumblrAPItarget.format('lolcats-lol-cat'))\n",
    "\n",
    "print(r.text[:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This might not look very good on first inspection, but it has far fewer angle\n",
    "braces than html, which makes it easier to parse. What we have is\n",
    "[JSON](https://en.wikipedia.org/wiki/JSON) a 'human readable' text based data\n",
    "transmission format based on javascript. Luckily, we can readily convert it to a\n",
    "python `dict`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We need to load only the stuff between the curly braces\n",
    "d = json.loads(r.text[len('var tumblr_api_read = '):-2])\n",
    "print(d.keys())\n",
    "print(len(d['posts']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we read the [API specification](https://www.tumblr.com/docs/en/api/v1), we\n",
    "will see there are a lot of things we can get if we add things to our GET\n",
    "request. First we can retrieve posts by their id number. Let's first get post\n",
    "`146020177084`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(tumblrAPItarget.format('lolcats-lol-cat'), params = {'id' : 146020177084})\n",
    "d = json.loads(r.text[len('var tumblr_api_read = '):-2])\n",
    "d['posts'][0].keys()\n",
    "d['posts'][0]['photo-url-1280']\n",
    "\n",
    "with open('lolcat.gif', 'wb') as f:\n",
    "    gifRequest = requests.get(d['posts'][0]['photo-url-1280'], stream = True)\n",
    "    f.write(gifRequest.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='lolcat.gif'>\n",
    "\n",
    "Such beauty; such vigor (If you can't see it you have to refresh the page). Now\n",
    "we could retrieve the text from all posts as well\n",
    "as related metadata, like the post date, caption or tags. We could also get\n",
    "links to all the images."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Putting a max in case the blog has millions of images\n",
    "#The given max will be rounded up to the nearest multiple of 50\n",
    "def tumblrImageScrape(blogName, maxImages = 200):\n",
    "    #Restating this here so the function isn't dependent on any external variables\n",
    "    tumblrAPItarget = 'http://{}.tumblr.com/api/read/json'\n",
    "\n",
    "    #There are a bunch of possible locations for the photo url\n",
    "    possiblePhotoSuffixes = [1280, 500, 400, 250, 100]\n",
    "\n",
    "    #These are the pieces of information we will be gathering,\n",
    "    #at the end we will convert this to a DataFrame.\n",
    "    #There are a few other datums we could gather like the captions\n",
    "    #you can read the Tumblr documentation to learn how to get them\n",
    "    #https://www.tumblr.com/docs/en/api/v1\n",
    "    postsData = {\n",
    "        'id' : [],\n",
    "        'photo-url' : [],\n",
    "        'date' : [],\n",
    "        'tags' : [],\n",
    "        'photo-type' : []\n",
    "    }\n",
    "\n",
    "    #Tumblr limits us to a max of 50 posts per request\n",
    "    for requestNum in range(maxImages // 50):\n",
    "        requestParams = {\n",
    "            'start' : requestNum * 50,\n",
    "            'num' : 50,\n",
    "            'type' : 'photo'\n",
    "        }\n",
    "        r = requests.get(tumblrAPItarget.format(blogName), params = requestParams)\n",
    "        requestDict = json.loads(r.text[len('var tumblr_api_read = '):-2])\n",
    "        for postDict in requestDict['posts']:\n",
    "            #We are dealing with uncleaned data, we can't trust it.\n",
    "            #Specifically, not all posts are guaranteed to have the fields we want\n",
    "            try:\n",
    "                postsData['id'].append(postDict['id'])\n",
    "                postsData['date'].append(postDict['date'])\n",
    "                postsData['tags'].append(postDict['tags'])\n",
    "            except KeyError as e:\n",
    "                raise KeyError(\"Post {} from {} is missing: {}\".format(postDict['id'], blogName, e))\n",
    "\n",
    "            foundSuffix = False\n",
    "            for suffix in possiblePhotoSuffixes:\n",
    "                try:\n",
    "                    photoURL = postDict['photo-url-{}'.format(suffix)]\n",
    "                    postsData['photo-url'].append(photoURL)\n",
    "                    postsData['photo-type'].append(photoURL.split('.')[-1])\n",
    "                    foundSuffix = True\n",
    "                    break\n",
    "                except KeyError:\n",
    "                    pass\n",
    "            if not foundSuffix:\n",
    "                #Make sure your error messages are useful\n",
    "                #You will be one of the users\n",
    "                raise KeyError(\"Post {} from {} is missing a photo url\".format(postDict['id'], blogName))\n",
    "\n",
    "    return pandas.DataFrame(postsData)\n",
    "tumblrImageScrape('lolcats-lol-cat', 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have the urls of a bunch of images and can run OCR on them to gather\n",
    "compelling meme narratives, accompanied by cats.\n",
    "\n",
    "# Files\n",
    "\n",
    "What if the text we want isn't on a webpage? There are a many other sources of\n",
    "text available, typically organized into *files*.\n",
    "\n",
    "## Raw text (and encoding)\n",
    "\n",
    "The most basic form of storing text is as a _raw text_ document. Source code\n",
    "(`.py`, `.r`, etc) is usually raw text as are text files (`.txt`) and those with\n",
    "many other extension (e.g., .csv, .dat, etc.). Opening an unknown file with a\n",
    "text editor is often a great way of learning what the file is.\n",
    "\n",
    "We can create a text file in python with the `open()` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#example_text_file = 'sometextfile.txt'\n",
    "#stringToWrite = 'A line\\nAnother line\\nA line with a few unusual symbols \\u2421 \\u241B \\u20A0 \\u20A1 \\u20A2 \\u20A3 \\u0D60\\n'\n",
    "stringToWrite = 'A line\\nAnother line\\nA line with a few unusual symbols ␡ ␛ ₠ ₡ ₢ ₣ ൠ\\n'\n",
    "\n",
    "with open(example_text_file, mode = 'w', encoding='utf-8') as f:\n",
    "    f.write(stringToWrite)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice the `encoding='utf-8'` argument, which specifies how we map the bits from\n",
    "the file to the glyphs (and whitespace characters like tab (`'\\t'`) or newline\n",
    "(`'\\n'`)) on the screen. When dealing only with latin letters, arabic numerals\n",
    "and the other symbols on America keyboards you usually do not have to worry\n",
    "about encodings as the ones used today are backwards compatible with\n",
    "[ASCII](https://en.wikipedia.org/wiki/ASCII), which gives the binary\n",
    "representation of 128 characters.\n",
    "\n",
    "Some of you, however, will want to use other characters (e.g., Chinese\n",
    "characters). To solve this there is\n",
    "[Unicode](https://en.wikipedia.org/wiki/Unicode) which assigns numbers to\n",
    "symbols, e.g., 041 is `'A'` and 03A3 is `'Σ'` (numbers starting with 0 are\n",
    "hexadecimal). Often non/beyond-ASCII characters are called Unicode characters.\n",
    "Unicode contains 1,114,112 characters, about 10\\% of which have been assigned.\n",
    "Unfortunately there are many ways used to map combinations of bits to Unicode\n",
    "symbols. The ones you are likely to encounter are called by Python _utf-8_,\n",
    "_utf-16_ and _latin-1_. _utf-8_ is the standard for Linux and Mac OS while both\n",
    "_utf-16_ and _latin-1_ are used by windows. If you use the wrong encoding,\n",
    "characters can appear wrong, sometimes change in number or Python could raise an\n",
    "exception. Lets see what happens when we open the file we just created with\n",
    "different encodings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(example_text_file, encoding='utf-8') as f:\n",
    "    print(\"This is with the correct encoding:\")\n",
    "    print(f.read())\n",
    "\n",
    "with open(example_text_file, encoding='latin-1') as f:\n",
    "    print(\"This is with the wrong encoding:\")\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that with _latin-1_ the unicode characters are mixed up and there are too\n",
    "many of them. You need to keep in mind encoding when obtaining text files.\n",
    "Determining the encoding can sometime involve substantial work."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also load many text files at once. LEts tart by looking at the Shakespeare files in the `data` directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/Shakespeare/midsummer_nights_dream.txt') as f:\n",
    "    midsummer = f.read()\n",
    "print(midsummer[-700:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then to load all the files in `../data/Shakespeare` we can use a for loop with `scandir`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "targetDir = '../data/Shakespeare' #Change this to your own directory of texts\n",
    "shakespearText = []\n",
    "shakespearFileName = []\n",
    "\n",
    "for file in (file for file in os.scandir(targetDir) if file.is_file() and not file.name.startswith('.')):\n",
    "    with open(file.path) as f:\n",
    "        shakespearText.append(f.read())\n",
    "    shakespearFileName.append(file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we can put them all in pandas DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shakespear_df = pandas.DataFrame({'text' : shakespearText}, index = shakespearFileName)\n",
    "shakespear_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting your text in a format like this is the first step of most analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PDF\n",
    "\n",
    "Another common way text will be stored is in a PDF file. First we will download\n",
    "a pdf in Python. To do that lets grab a chapter from\n",
    "_Speech and Language Processing_, chapter 21 is on Information Extraction which\n",
    "seems apt. It is stored as a pdf at [https://web.stanford.edu/~jurafsky/slp3/21.\n",
    "pdf](https://web.stanford.edu/~jurafsky/slp3/21.pdf) although we are downloading\n",
    "from a copy just in case Jurafsky changes their website."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#information_extraction_pdf = 'https://github.com/KnowledgeLab/content_analysis/raw/data/21.pdf'\n",
    "\n",
    "infoExtractionRequest = requests.get(information_extraction_pdf, stream=True)\n",
    "print(infoExtractionRequest.text[:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It says `'pdf'`, so thats a good sign. The rest though looks like we are having\n",
    "issues with an encoding. The random characters are not caused by our encoding\n",
    "being wrong, however. They are cause by there not being an encoding for those\n",
    "parts at all. PDFs are nominally binary files, meaning there are sections of\n",
    "binary that are specific to pdf and nothing else so you need something that\n",
    "knows about pdf to read them. To do that we will be using\n",
    "[`PyPDF2`](https://github.com/mstamy2/PyPDF2), a PDF processing library for\n",
    "Python 3.\n",
    "\n",
    "\n",
    "Because PDFs are a very complicated file format pdfminer requires a large amount\n",
    "of boilerplate code to extract text, we have written a function that takes in an\n",
    "open PDF file and returns the text so you don't have to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readPDF(pdfFile):\n",
    "    #Based on code from http://stackoverflow.com/a/20905381/4955164\n",
    "    #Using utf-8, if there are a bunch of random symbols try changing this\n",
    "    codec = 'utf-8'\n",
    "    rsrcmgr = pdfminer.pdfinterp.PDFResourceManager()\n",
    "    retstr = io.StringIO()\n",
    "    layoutParams = pdfminer.layout.LAParams()\n",
    "    device = pdfminer.converter.TextConverter(rsrcmgr, retstr, laparams = layoutParams, codec = codec)\n",
    "    #We need a device and an interpreter\n",
    "    interpreter = pdfminer.pdfinterp.PDFPageInterpreter(rsrcmgr, device)\n",
    "    password = ''\n",
    "    maxpages = 0\n",
    "    caching = True\n",
    "    pagenos=set()\n",
    "    for page in pdfminer.pdfpage.PDFPage.get_pages(pdfFile, pagenos, maxpages=maxpages, password=password,caching=caching, check_extractable=True):\n",
    "        interpreter.process_page(page)\n",
    "    device.close()\n",
    "    returnedString = retstr.getvalue()\n",
    "    retstr.close()\n",
    "    return returnedString"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we need to take the response object and convert it into a 'file like'\n",
    "object so that pdfminer can read it. To do this we will use `io`'s `BytesIO`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "infoExtractionBytes = io.BytesIO(infoExtractionRequest.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can give it to pdfminer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(readPDF(infoExtractionBytes)[:550])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From here we can either look at the full text or fiddle with our PDF reader and\n",
    "get more information about individual blocks of text.\n",
    "\n",
    "## Word Docs\n",
    "\n",
    "The other type of document you are likely to encounter is the `.docx`, these are\n",
    "actually a version of [XML](https://en.wikipedia.org/wiki/Office_Open_XML), just\n",
    "like HTML, and like HTML we will use a specialized parser.\n",
    "\n",
    "For this class we will use [`python-docx`](https://python-\n",
    "docx.readthedocs.io/en/latest/) which provides a nice simple interface for\n",
    "reading `.docx` files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#example_docx = 'https://github.com/KnowledgeLab/content_analysis/raw/data/example_doc.docx'\n",
    "\n",
    "r = requests.get(example_docx, stream=True)\n",
    "d = docx.Document(io.BytesIO(r.content))\n",
    "for paragraph in d.paragraphs[:7]:\n",
    "    print(paragraph.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This procedure uses the `io.BytesIO` class again, since `docx.Document` expects\n",
    "a file. Another way to do it is to save the document to a file and then read it\n",
    "like any other file. If we do this we can either delete the file afterwords, or\n",
    "save it and avoid downloading the following time.\n",
    "\n",
    "This function is useful as a part of many different tasks so it and others like it will be added to the helper package `lucem_illud_2020` so we can use it later without having to retype it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def downloadIfNeeded(targetURL, outputFile, **openkwargs):\n",
    "    if not os.path.isfile(outputFile):\n",
    "        outputDir = os.path.dirname(outputFile)\n",
    "        #This function is a more general os.mkdir()\n",
    "        if len(outputDir) > 0:\n",
    "            os.makedirs(outputDir, exist_ok = True)\n",
    "        r = requests.get(targetURL, stream=True)\n",
    "        #Using a closure like this is generally better than having to\n",
    "        #remember to close the file. There are ways to make this function\n",
    "        #work as a closure too\n",
    "        with open(outputFile, 'wb') as f:\n",
    "            f.write(r.content)\n",
    "    return open(outputFile, **openkwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function will download, save and open `outputFile` as `outputFile` or just\n",
    "open it if `outputFile` exists. By default `open()` will open the file as read\n",
    "only text with the local encoding, which may cause issues if its not a text\n",
    "file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    d = docx.Document(downloadIfNeeded(example_docx, example_docx_save))\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to tell `open()` to read in binary mode (`'rb'`), this is why we added\n",
    "`**openkwargs`, this allows us to pass any keyword arguments (kwargs) from\n",
    "`downloadIfNeeded` to `open()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = docx.Document(downloadIfNeeded(example_docx, example_docx_save, mode = 'rb'))\n",
    "for paragraph in d.paragraphs[:7]:\n",
    "    print(paragraph.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can read the file with `docx.Document` and not have to wait for it to be\n",
    "downloaded every time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# <span style=\"color:red\">Section 3</span>\n",
    "<span style=\"color:red\">Construct cells immediately below this that extract and organize textual content from text, PDF or Word into a pandas dataframe.</span>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I'm first going to extract all of the PMIDS from the text file I've kept them into a list, and then a pandas data frame. This isn't strictly text analysis, I know, but it's for my project and technically every item in the CSV is a string. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pmids_list = []\n",
    "with open('pmids_na_omit.txt') as csvfile:\n",
    "    pmids = csv.reader(csvfile, delimiter=',')\n",
    "    for row in pmids:\n",
    "        for i in row: \n",
    "            pmids_list.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30784590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30692680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30659290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30609404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30445657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4035</th>\n",
       "      <td>28468917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4036</th>\n",
       "      <td>28414085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4037</th>\n",
       "      <td>28408976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4038</th>\n",
       "      <td>28358245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4039</th>\n",
       "      <td>28087420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4040 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0\n",
       "0     30784590\n",
       "1     30692680\n",
       "2     30659290\n",
       "3     30609404\n",
       "4     30445657\n",
       "...        ...\n",
       "4035  28468917\n",
       "4036  28414085\n",
       "4037  28408976\n",
       "4038  28358245\n",
       "4039  28087420\n",
       "\n",
       "[4040 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pmids_df = pd.DataFrame(pmids_list)\n",
    "pmids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.05 ms, sys: 4.07 ms, total: 8.12 ms\n",
      "Wall time: 10.9 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from pubmed_lookup import PubMedLookup\n",
    "from pubmed_lookup import Publication\n",
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pubmed_lookup/pubmed_lookup.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, pubmed_record, resolve_doi)\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0missue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Issue'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpages\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Pages'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 47\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_article_url\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresolve_doi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mresolve_doi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     48\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m         \u001b[0mxml_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_pubmed_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pubmed_lookup/pubmed_lookup.py\u001b[0m in \u001b[0;36mset_article_url\u001b[0;34m(self, resolve_doi)\u001b[0m\n\u001b[1;32m    189\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresolve_doi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    190\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 191\u001b[0;31m                     \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdoi_url\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    192\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mConnectionResetError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mURLError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    193\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0murl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdoi_url\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    220\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    221\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 222\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    223\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0minstall_opener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopener\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    529\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mprocessor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess_response\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    530\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 531\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    532\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    533\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    638\u001b[0m         \u001b[0;31m# request was successfully received, understood, and accepted.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    639\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 640\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    641\u001b[0m                 'http', request, response, code, msg, hdrs)\n\u001b[1;32m    642\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    561\u001b[0m             \u001b[0mhttp_err\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    562\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 563\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    564\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    565\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    500\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    501\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 502\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    503\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_302\u001b[0;34m(self, req, fp, code, msg, headers)\u001b[0m\n\u001b[1;32m    753\u001b[0m         \u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    754\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 755\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    756\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    757\u001b[0m     \u001b[0mhttp_error_301\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_303\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_307\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_302\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    523\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    524\u001b[0m         \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maudit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'urllib.Request'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 525\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    526\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    527\u001b[0m         \u001b[0;31m# post-process response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36m_open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    540\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    541\u001b[0m         \u001b[0mprotocol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 542\u001b[0;31m         result = self._call_chain(self.handle_open, protocol, protocol +\n\u001b[0m\u001b[1;32m    543\u001b[0m                                   '_open', req)\n\u001b[1;32m    544\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    500\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    501\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 502\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    503\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mhttps_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1358\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1359\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mhttps_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1360\u001b[0;31m             return self.do_open(http.client.HTTPSConnection, req,\n\u001b[0m\u001b[1;32m   1361\u001b[0m                 context=self._context, check_hostname=self._check_hostname)\n\u001b[1;32m   1362\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py\u001b[0m in \u001b[0;36mdo_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1319\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# timeout error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1320\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mURLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1321\u001b[0;31m             \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1322\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1323\u001b[0m             \u001b[0mh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1321\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1322\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1323\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1324\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;31m# read until we get a non-100 response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 303\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    304\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 264\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    265\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    266\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mLineTooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"status line\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 669\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    670\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1239\u001b[0m                   \u001b[0;34m\"non-zero flags not allowed in calls to recv_into() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1240\u001b[0m                   self.__class__)\n\u001b[0;32m-> 1241\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1242\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1243\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1097\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbuffer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1099\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1100\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1101\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "%%time\n",
    "abs_corpus = []\n",
    "for link in all_links: \n",
    "    try: \n",
    "        email = 'deblina@uchicago.edu'\n",
    "        lookup = PubMedLookup(link, email)\n",
    "        publication = Publication(lookup) \n",
    "        abs_corpus.append(publication)\n",
    "    except TypeError: \n",
    "        pass \n",
    "    except RuntimeError: \n",
    "        pass "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<pubmed_lookup.pubmed_lookup.Publication at 0x110e4e9a0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x110eb7af0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264b2610>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126514760>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x110ecd250>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264a2be0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126512c70>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12622baf0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12622bfd0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12653f6d0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264ffee0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264ff8e0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264a2ee0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126590e80>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12622beb0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126512790>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8eb0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x110eb7d90>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a6dc0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8ca0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8f40>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8af0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8a30>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265ebf40>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265a8760>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265eba90>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266176a0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126617fa0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265ebaf0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265ebcd0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126617ca0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265e65b0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265eb310>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265e68b0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265eb850>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126649e80>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265e6a60>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266226a0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126622c40>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126622be0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126649df0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126649a90>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126668700>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1265e6df0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126688850>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266880d0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1264ff0a0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126682dc0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126514fd0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12657c280>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126636e50>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126682c40>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126636fd0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126682f40>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266c0730>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266adf70>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266c0820>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266c0bb0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266c0940>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126688dc0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266881c0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12671ceb0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12672f7f0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12674ca00>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x12674c400>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266d7eb0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266d7ee0>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126739b20>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x1266d7a30>,\n",
       " <pubmed_lookup.pubmed_lookup.Publication at 0x126758e50>]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abs_corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Identifying Extrinsic versus Intrinsic Drivers of Variation in Cell Behavior in Human iPSC Lines from Healthy Donors.': ['Large cohorts of human induced pluripotent stem cells (iPSCs) from healthy donors are a potentially powerful tool for investigating the relationship between genetic variants and cellular behavior. Here, we integrate high content imaging of cell shape, proliferation, and other phenotypes with gene expression and DNA sequence datasets from over 100 human iPSC lines. By applying a dimensionality reduction approach, Probabilistic Estimation of Expression Residuals (PEER), we extracted factors that captured the effects of intrinsic (genetic concordance between different cell lines from the same donor) and extrinsic (cell responses to different fibronectin concentrations) conditions. We identify genes that correlate in expression with intrinsic and extrinsic PEER factors and associate outlier cell behavior with genes containing rare deleterious non-synonymous SNVs. Our study, thus, establishes a strategy for examining the genetic basis of inter-individual variability in cell behavior.',\n",
       "  '30784590'],\n",
       " 'GARFIELD classifies disease-relevant genomic features through integration of functional annotations with association signals.': [\"Loci discovered by genome-wide association studies predominantly map outside protein-coding genes. The interpretation of the functional consequences of non-coding variants can be greatly enhanced by catalogs of regulatory genomic regions in cell lines and primary tissues. However, robust and readily applicable methods are still lacking by which to systematically evaluate the contribution of these regions to genetic variation implicated in diseases or quantitative traits. Here we propose a novel approach that leverages genome-wide association studies' findings with regulatory or functional annotations to classify features relevant to a phenotype of interest. Within our framework, we account for major sources of confounding not offered by current methods. We further assess enrichment of genome-wide association studies for 19 traits within Encyclopedia of DNA Elements- and Roadmap-derived regulatory regions. We characterize unique enrichment patterns for traits and annotations driving novel biological insights. The method is implemented in standalone software and an R package, to facilitate its application by the research community.\",\n",
       "  '30692680'],\n",
       " 'Author Correction: Landscape of somatic mutations in 560 breast cancer whole-genome sequences.': [\"In the Methods section of this Article, 'greater than' should have been 'less than' in the sentence 'Putative regions of clustered rearrangements were identified as having an average inter-rearrangement distance that was at least 10 times greater than the whole-genome average for the individual sample.\\u200a'. The Article has not been corrected.\",\n",
       "  '30659290'],\n",
       " 'Integrating Genomics into Healthcare: A Global Responsibility.': ['Genomic sequencing is rapidly transitioning into clinical practice, and implementation into healthcare systems has been supported by substantial government investment, totaling over US$4 billion, in at least 14 countries. These national genomic-medicine initiatives are driving transformative change under real-life conditions while simultaneously addressing barriers to implementation and gathering evidence for wider adoption. We review the diversity of approaches and current progress made by national genomic-medicine initiatives in the UK, France, Australia, and US and provide a roadmap for sharing strategies, standards, and data internationally to accelerate implementation.',\n",
       "  '30609404'],\n",
       " 'The European Bioinformatics Institute in 2018: tools, infrastructure and training.': ['The European Bioinformatics Institute (https://www.ebi.ac.uk/) archives, curates and analyses life sciences data produced by researchers throughout the world, and makes these data available for re-use globally (https://www.ebi.ac.uk/). Data volumes continue to grow exponentially: total raw storage capacity now exceeds 160 petabytes, and we manage these increasing data flows while maintaining the quality of our services. This year we have improved the efficiency of our computational infrastructure and doubled the bandwidth of our connection to the worldwide web. We report two new data resources, the Single Cell Expression Atlas (https://www.ebi.ac.uk/gxa/sc/), which is a component of the Expression Atlas; and the PDBe-Knowledgebase (https://www.ebi.ac.uk/pdbe/pdbe-kb), which collates functional annotations and predictions for structure data in the Protein Data Bank. Additionally, Europe PMC (http://europepmc.org/) has added preprint abstracts to its search results, supplementing results from peer-reviewed publications. EMBL-EBI maintains over 150 analytical bioinformatics tools that complement our data resources. We make these tools available for users through a web interface as well as programmatically using application programming interfaces, whilst ensuring the latest versions are available for our users. Our training team, with support from all of our staff, continued to provide on-site, off-site and web-based training opportunities for thousands of researchers worldwide this year.',\n",
       "  '30445657'],\n",
       " 'Inability of bats to synthesise L-ascorbic acid.': ['', '1264230'],\n",
       " 'Dental ontogeny and adaptation in Diphylla ecaudata.': ['', '1113041'],\n",
       " 'Effects of sodium chloride on water consumption, weight, and survival in the woodrats, Neotoma micropus and Neotoma floridana.': ['',\n",
       "  '5425661'],\n",
       " 'Notes on distribution and reproduction of some bats from Kansas, with remarks on incidence of rabies.': ['',\n",
       "  '5582959'],\n",
       " 'High-Throughput Sequencing in Respiratory, Critical Care, and Sleep Medicine Research. An Official American Thoracic Society Workshop Report.': ['High-throughput, \"next-generation\" sequencing methods are now being broadly applied across all fields of biomedical research, including respiratory disease, critical care, and sleep medicine. Although there are numerous review articles and best practice guidelines related to sequencing methods and data analysis, there are fewer resources summarizing issues related to study design and interpretation, especially as applied to common, complex, nonmalignant diseases. To address these gaps, a single-day workshop was held at the American Thoracic Society meeting in May 2017, led by the American Thoracic Society Section on Genetics and Genomics. The aim of this workshop was to review the design, analysis, interpretation, and functional follow-up of high-throughput sequencing studies in respiratory, critical care, and sleep medicine research. This workshop brought together experts in multiple fields, including genetic epidemiology, biobanking, bioinformatics, and research ethics, along with physician-scientists with expertise in a range of relevant diseases. The workshop focused on application of DNA and RNA sequencing research in common chronic diseases and did not cover sequencing studies in lung cancer, monogenic diseases (e.g., cystic fibrosis), or microbiome sequencing. Participants reviewed and discussed study design, data analysis and presentation, interpretation, functional follow-up, and reporting of results. This report summarizes the main conclusions of the workshop, specifically addressing the application of these methods in respiratory, critical care, and sleep medicine research. This workshop report may serve as a resource for our research community as well as for journal editors and reviewers of sequencing-based manuscript submissions in our research field.',\n",
       "  '30592451'],\n",
       " 'Gene Regulatory Network Analysis Identifies Sex-Linked Differences in Colon Cancer Drug Metabolism.': ['Understanding sex differences in colon cancer is essential to advance disease prevention, diagnosis, and treatment. Males have a higher risk of developing colon cancer and a lower survival rate than women. However, the molecular features that drive these sex differences are poorly understood. In this study, we use both transcript-based and gene regulatory network methods to analyze RNA-seq data from The Cancer Genome Atlas for 445 patients with colon cancer. We compared gene expression between tumors in men and women and observed significant sex differences in sex chromosome genes only. We then inferred patient-specific gene regulatory networks and found significant regulatory differences between males and females, with drug and xenobiotics metabolism via cytochrome P450 pathways more strongly targeted in females. This finding was validated in a dataset of 1,193 patients from five independent studies. While targeting, the drug metabolism pathway did not change overall survival for males treated with adjuvant chemotherapy, females with greater targeting showed an increase in 10-year overall survival probability, 89% [95% confidence interval (CI), 78-100] survival compared with 61% (95% CI, 45-82) for women with lower targeting, respectively ( = 0.034). Our network analysis uncovers patterns of transcriptional regulation that differentiate male and female colon cancer and identifies differences in regulatory processes involving the drug metabolism pathway associated with survival in women who receive adjuvant chemotherapy. This approach can be used to investigate the molecular features that drive sex differences in other cancers and complex diseases. A network-based approach reveals that sex-specific patterns of gene targeting by transcriptional regulators are associated with survival outcome in colon cancer. This approach can be used to understand how sex influences progression and response to therapies in other cancers. .',\n",
       "  '30275053'],\n",
       " 'Histopathological Image QTL Discovery of Immune Infiltration Variants.': ['Genotype-to-phenotype association studies typically use macroscopic physiological measurements or molecular readouts as quantitative traits. There are comparatively few suitable quantitative traits available between cell and tissue length scales, a limitation that hinders our ability to identify variants affecting phenotype at many clinically informative levels. Here we show that quantitative image features, automatically extracted from histopathological imaging data, can be used for image quantitative trait loci (iQTLs) mapping and variant discovery. Using thyroid pathology images, clinical metadata, and genomics data from the Genotype-Tissue Expression (GTEx) project, we establish and validate a quantitative imaging biomarker for immune cell infiltration. A total of 100,215 variants were selected for iQTL profiling and tested for genotype-phenotype associations with our quantitative imaging biomarker. Significant associations were found in HDAC9 and TXNDC5. We validated the TXNDC5 association using GTEx cis-expression QTL data and an independent hypothyroidism dataset from the Electronic Medical Records and Genomics network.',\n",
       "  '30240647'],\n",
       " 'Artificial intelligence in radiology.': ['Artificial intelligence (AI) algorithms, particularly deep learning, have demonstrated remarkable progress in image-recognition tasks. Methods ranging from convolutional neural networks to variational autoencoders have found myriad applications in the medical image analysis field, propelling it forward at a rapid pace. Historically, in radiology practice, trained physicians visually assessed medical images for the detection, characterization and monitoring of diseases. AI methods excel at automatically recognizing complex patterns in imaging data and providing quantitative, rather than qualitative, assessments of radiographic characteristics. In this Opinion article, we establish a general understanding of AI methods, particularly those pertaining to image-based tasks. We explore how these methods could impact multiple facets of radiology, with a general focus on applications in oncology, and demonstrate ways in which these methods are advancing the field. Finally, we discuss the challenges facing clinical implementation and provide our perspective on how the domain could be advanced.',\n",
       "  '29777175'],\n",
       " 'Testing CP and CPT violation in the neutral kaon system at a phi factory.': ['',\n",
       "  '10014315'],\n",
       " 'Chiral anomalies in two-dimensional quantum field theory.': ['', '10011709'],\n",
       " 'The death of a pet. How it can affect owners.': [\"The death of a family pet can be a difficult and upsetting experience. In this article, human reactions after pet death are described as a normal bereavement sequence and a case study is used to describe and explain major issues associated with a pet's death.\",\n",
       "  '3872519'],\n",
       " 'Helping people adjust to the death of a pet.': [\"Recently the bond that exists between people and their pets and its impact on physical and mental health have become topics of interest. This descriptive study examines the provision of social work services in a veterinary hospital to individuals who are distressed about a pet's illness or death and considers the implications of this kind of service delivery for social work.\",\n",
       "  '6706255'],\n",
       " 'The role of the residential treatment center in the treatment of children and adolescents with emotional problems.': ['',\n",
       "  '5089140'],\n",
       " 'The current practice and care of paediatric patients post cardiac catheterisation.': ['BACKGROUND: Literature is lacking to guide standardised care and assessment practices for paediatric patients post cardiac catheterisation. In response to this gap, we sought to describe the current state of practice in cardiology programmes performing paediatric cardiac catheterisations procedures in the United States of America.Materials and methodsA web-based survey was distributed to the Congenital Cardiovascular Interventional Study Consortium Listserv, with representation from 113 identified institutions. A 36-question survey, including fixed-choice and open-ended questions, was developed and piloted for reliability and validity before distribution. Data were summarised descriptively with count and frequency or median and range.\\n\\nRESULTS: Of the 113 identified institutions, 52% (n=59) responded to the survey. Manual pressure is used to achieve haemostasis by 94.9% of the respondents. Pressure dressings are used by a majority of the facilities and the length of time for bed rest is variable, with the majority using 6 hours for arterial access and 4 hours for venous access. Predominantly, respondents use the time of haemostasis as the start time of bed rest while a third of respondents reported using the time the sheath was removed.\\n\\nCONCLUSION: In this study, variation in a number of post catheterisation care and assessment practices for paediatric patients was noted across cardiology programmes. Information from this assessment identifies key opportunities to collaborate in developing standardised practices for the care and assessment of the paediatric patients post catheterisation.',\n",
       "  '30474576'],\n",
       " 'Current pharmacological treatment guidelines for psoriasis and psoriatic arthritis.': ['INTRODUCTION: Psoriasis is a common chronic skin condition that is prevalent worldwide. Despite the numerous treatment options available, discrepancies exist between international guidelines. Areas covered: This review aims to evaluate the current international guidelines for the systemic treatment of psoriasis and psoriatic arthritis. Expert commentary: Most guidelines are unanimous on medication dosing and laboratory monitoring. However, even the most up-to-date guidelines do not include many of the new biologic medications. Guidelines will require frequent updates to include the newer treatments and will soon need to state a recommendation on the use of biosimilars.',\n",
       "  '30449211'],\n",
       " 'Full-field imaging of thermal and acoustic dynamics in an individual nanostructure using tabletop high harmonic beams.': ['Imaging charge, spin, and energy flow in materials is a current grand challenge that is relevant to a host of nanoenhanced systems, including thermoelectric, photovoltaic, electronic, and spin devices. Ultrafast coherent x-ray sources enable functional imaging on nanometer length and femtosecond timescales particularly when combined with advances in coherent imaging techniques. Here, we combine ptychographic coherent diffractive imaging with an extreme ultraviolet high harmonic light source to directly visualize the complex thermal and acoustic response of an individual nanoscale antenna after impulsive heating by a femtosecond laser. We directly image the deformations induced in both the nickel tapered nanoantenna and the silicon substrate and see the lowest-order generalized Lamb wave that is partially confined to a uniform nanoantenna. The resolution achieved-sub-100 nm transverse and 0.5-Å axial spatial resolution, combined with ≈10-fs temporal resolution-represents a significant advance in full-field dynamic imaging capabilities. The tapered nanoantenna is sufficiently complex that a full simulation of the dynamic response would require enormous computational power. We therefore use our data to benchmark approximate models and achieve excellent agreement between theory and experiment. In the future, this work will enable three-dimensional functional imaging of opaque materials and nanostructures that are sufficiently complex that their functional properties cannot be predicted.',\n",
       "  '30345364'],\n",
       " 'Predictive genetic profiles for regional lymph node metastasis in primary cutaneous melanoma: a case-matched pilot study.': ['Melanoma confers an estimated lifetime risk of one in 50 for 2016. Clinicopathologic staging and sentinel lymph node biopsy (SLNB) have been the standard of care for T2 and T3 lesions. Molecular biomarkers identified in the primary lesion suggestive of metastatic potential may offer a more conclusive prognosis of these lesions. Our purpose was to investigate molecular mutations in primary melanoma that were predictive for micrometastasis as defined by a positive sentinel lymph node (SLN) in a case-controlled manner: nine patients with negative SLN and nine with positive SLN. The two cohorts were statistically identical as shown by a t-test for age (P=0.17), race (P=0.18), Breslow depth (P=0.14), Clark level (P=0.33), host response (P=0.17), ulceration (P=0.50), satellite nodules (P=0.17), lymphovascular invasion (P=0.50), and mitotic activity (P=0.09). While no single gene was significantly associated with SLN status, multivariate analysis using classification and regression tree assessment revealed two unique gene profiles that completely represented regional metastases in our cohort as defined by a positive SLN: PIK3CA (+) NRAS (-) and PIK3CA (-) ERBB4 (-) TP53 (+) SMAD4 (-). These profiles were identified in 89% of the patients with positive SLN; none of these profiles were identified in the SLN-negative cohort. We identified two unique gene profiles associated with positive SLN that do not overlap other studies and highlight the genetic complexity that portends the metastatic phenotype in cutaneous melanoma.',\n",
       "  '30179987'],\n",
       " 'Early life trauma: An exploratory study of effects on OXTR and NR3C1 gene expression and nurturing self-efficacy in mothers of infants.': ['BACKGROUND: In animals, adverse early experience alters oxytocinergic and glucocorticoid activity and maternal behavior in adulthood. This preliminary study explored associations among childhood trauma (loss of a parent or sexual abuse in childhood), maternal self-efficacy, and leukocyte gene expression (mRNA) of oxytocin and glucocorticoid receptors (OXTR and NR3C1) in mothers of infants.\\n\\nMETHODS: 62 mothers (20 with early life trauma) with healthy 3-month old infants reported maternal self-efficacy, depression, infant temperament, and overall social support; the effects of early trauma on these measures were assessed. Of these, 35 mothers (14 with early trauma) underwent blood draws after 2 infant feeding times; their OXTR and NR3C1 mRNA was compared to a control group of 25 no-infant women without early trauma, and also was examined for associations with self-efficacy.\\n\\nRESULTS: OXTR mRNA was increased in mothers of infants versus no-infant controls (p\\u202f<\\u202f0.0003), and mothers with greatest prior maternal experience had higher OXTR than those with less experience (0-2 vs. 3+ older children, p\\u202f<\\u202f0.033). Mothers with early trauma and less maternal experience had lower OXTR mRNA than no-trauma mothers (p\\u202f<\\u202f0.029) and lower NR3C1 mRNA than controls (p\\u202f<\\u202f0.004). Mothers with depression also had lower NR3C1 than other mothers (p\\u202f<\\u202f0.003) but did not differ in OXTR. Mothers with early trauma also reported their support network to be less helpful and more upsetting and unpredictable than other mothers (p\\u202f<\\u202f0.035-p\\u202f<\\u202f0.005). Regarding maternal behavior, in mothers with early trauma, helpful support networks increased self-reported nurturing self-efficacy when babies were not fussy but decreased it with fussy babies (p\\u202f<\\u202f0.05). Support was unrelated to self-efficacy in no-trauma mothers. Similarly, among mothers with low OXTR or NR3C1 (-1SD, most having early trauma and lower maternal experience), greater support was associated with lower self-efficacy (p\\u202f<\\u202f0.05), while mothers with high OXTR or NR3C1 (+1SD) tended to have higher self-efficacy with greater support.\\n\\nCONCLUSIONS: These preliminary findings need confirmation in a larger sample but suggest that childhood trauma influences maternal behavior and both OXTR and NR3C1 pathways in mothers of infants, and that both depression and prior maternal experience may be other important factors. Effects on maternal behavior appear to require more complex modeling.',\n",
       "  '29580902'],\n",
       " 'Nature of antibiosis manifested by selected species of the Mucorales.': ['',\n",
       "  '13425430'],\n",
       " 'Solvent considerations in fungicidal testing.': ['', '12990522'],\n",
       " 'Fungicidal and fungistatic evaluation of certain phenols and surface-active agents.': ['',\n",
       "  '14907503'],\n",
       " 'Antifungal properties of some antihistamines.': ['', '14861110'],\n",
       " 'Traumatized brains.': ['', '30926926'],\n",
       " 'Listening to vibrations.': ['', '30926925'],\n",
       " 'Double satiety.': ['', '30926924'],\n",
       " 'Flowing with the visual stream.': ['', '30926922'],\n",
       " 'The Evolution of the European Young Pediatricians Association (EURYPA).': ['',\n",
       "  '30922499'],\n",
       " 'The treatment of malocclusion with and without the removal of dental units.': ['',\n",
       "  '20996781'],\n",
       " 'THE PREPARATION AND PROPERTIES OF HIGHLY PURIFIED ASCORBIC ACID OXIDASE.': ['1. A method is described for the preparation of a highly purified ascorbic acid oxidase containing 0.24 per cent copper. 2. Using comparable activity measurements, this oxidase is about one and a half times as active on a dry weight basis as the hitherto most highly purified preparation described by Lovett-Janison and Nelson. The latter contained 0.15 per cent copper. 3. The oxidase activity is proportional to the copper content and the proportionality factor is the same as that reported by Lovett-Janison and Nelson. 4. When dialyzed free of salt, the blue concentrated oxidase solutions precipitate a dark green-blue protein which carries the activity. This may be prevented by keeping the concentrated solutions about 0.1 M in Na(2)HPO(4). 5. When highly diluted for activity measurements the oxidase rapidly loses activity (irreversibly) previous to the measurement, unless the dilution is made with a dilute inert protein (gelatin) solution. Therefore activity values obtained using such gelatin-stabilized dilute solutions of the oxidase run considerably higher than values obtained by the Lovett-Janison and Nelson technique. 6. The effect of pH and substrate concentration on the activity of the purified oxidase in the presence and absence of inert protein was studied.',\n",
       "  '19873382'],\n",
       " 'ENEMATA: FROM THE ANATOMICAL AND PHYSIOLOGICAL STANDPOINTS.': ['',\n",
       "  '18741625'],\n",
       " 'The assimilation of aluminium by the human system.': ['', '16744786'],\n",
       " 'Dentistry, Oral Prophylaxis and Bodily Health; Their Inter-relationship.': ['',\n",
       "  '20891193'],\n",
       " 'Detection of Gluten in Gluten-Free Labeled Restaurant Food: Analysis of Crowd-Sourced Data.': [\"INTRODUCTION: Adherence to a gluten-free (GF) diet is the mainstay of therapy for celiac disease. Until now, those wishing to avoid gluten in restaurants had to rely on menu labels, word of mouth, intuition, and restaurant workers' advice, with a relative dearth of supporting data. We used crowd-sourced data from users of a portable gluten detection device to estimate rates of, and identify risk factors for, gluten contamination of supposed GF restaurant foods.\\n\\nMETHODS: We analyzed data from a portable gluten detection device (Nima), collected across the United States during an 18-month period by users who opted to share the results of their point-of-care tests. Data were sorted by region, time of day, median household income in the restaurant's vicinity, restaurant genre, and food items. We used the χ test for bivariate analysis and multiple logistic regression for multivariate analysis to identify predictors of gluten detection in restaurant food.\\n\\nRESULTS: There were 5,624 tests, performed by 804 users, in the examined period. Gluten was detected in 32% of GF labeled foods. Rates of gluten detection differed by meal, with 27.2% at breakfast and 34.0% at dinner (P = 0.0008). GF labeled pizza and pasta were most likely to test positive for gluten, with gluten detected in 53.2% of pizza and 50.8% of pasta samples. On multivariate analysis, GF labeled food was less likely to test positive for gluten in the West than in the Northeast United States (odds ratio 0.80; 95% confidence interval 0.67-0.95).\\n\\nCONCLUSIONS: This study of crowd-sourced data suggests that a substantial fraction of GF labeled restaurant foods contain detectable gluten. Although the highly sensitive Nima device may detect gluten at levels <20 parts per million (ppm), leading to gluten exposure of unknown clinical significance, our findings raise a potential concern. In addition, our findings of higher rates of gluten detection in pizza and pasta provide practical data when providing dining strategies for patients with celiac disease.\",\n",
       "  '30920417'],\n",
       " 'Telomerase Impinges on the Cellular Response to Oxidative Stress Through Mitochondrial ROS-Mediated Regulation of Autophagy.': [\"Telomerase has cellular functions beyond telomere stabilization, including a role in mitochondria. The function of the catalytic component-TERT-in mitochondria is still unknown, but it seems to play a role in the response to oxidative stress. Here, we interrogated the role of the subcellular localization of TERT to the response to hydrogen peroxide (H₂O₂) treatment. Using normal human fibroblasts (NHF) expressing non-tagged wild type (WT) human TERT (hTERT) or nuclear localization and function (hTERT), a mutant that we previously described as being competent in telomere elongation, while not being able to localize to mitochondria, we found the differential activation of autophagy as a function of hTERT's subcellular localization. Specifically, we found that only cells expressing the mutant had significant increases in autophagy markers as a response to H₂O₂ challenge. Either the reintroduction of the mitochondrial pool of hTERT or the expression of mitochondrially-targeted catalase in mutant cells blunted the autophagic response under oxidative stress. Interestingly, autophagy activation was also associated with decreased levels of mitochondrial DNA damage. Taken together, these results suggest that the loss of hTERT in mitochondria initiates a signaling cascade that allows for cells to adapt to and cope with the lack of mitochondrial telomerase. Such effects also influence the cellular response to oxidative damage.\",\n",
       "  '30917518'],\n",
       " 'Adverse events in patients with high platelet reactivity following successful chronic total occlusion PCI: The Assessment of Dual AntiPlatelet Therapy with Drug-Eluting Stents (ADAPT-DES) study.': ['BACKGROUND: Chronic total occlusion (CTO) percutaneous coronary intervention (PCI) typically requires a greater number of stents and longer stent length than non-CTO PCI, placing these patients at greater risk for adverse ischemic events. We sought to determine whether the association between high platelet reactivity (HPR) and the risk of ischemic events is stronger after CTO than non-CTO PCI.\\n\\nMETHODS: Patients undergoing successful PCI in the multicenter ADAPT-DES study were stratified according to whether they underwent PCI of a CTO. HPR was defined as VerifyNow platelet reaction units >208. The study primary endpoint was the 2-year risk target vessel failure ([TVF] defined as cardiac death, myocardial infarction, or target lesion revascularization).\\n\\nRESULTS: CTO PCI was performed in 400 of 8448 patients. HPR was present in 34.5% of CTO PCI patients and 43.1% of non-CTO PCI patients (P = .0007). Patients undergoing CTO PCI with versus without HPR had significantly higher 2-year rates of TVF (15.0% versus 8.3%, P = .04) without significant differences in bleeding. HPR was an independent predictor of 2-year TVF (adjusted HR 1.16, 95% CI 1.02-1.34, P = .03) whereas CTO PCI was not (adjusted HR 0.89, 95% CI 0.65-1.22, P = .48). There was a significant interaction between CTO versus non-CTO PCI and PRU as a continuous variable for 2-year TVF (P = 0.02).\\n\\nCONCLUSIONS: In ADAPT-DES, HPR was associated with an increased 2-year risk of TVF after PCI, an association that was at least as strong after CTO PCI compared with non-CTO PCI.',\n",
       "  '30897527'],\n",
       " 'HTLV-1 Tax-1 interacts with SNX27 to regulate cellular localization of the HTLV-1 receptor molecule, GLUT1.': ['An estimated 10-20 million people worldwide are infected with human T cell leukemia virus type 1 (HTLV-1), with endemic areas of infection in Japan, Australia, the Caribbean, and Africa. HTLV-1 is the causative agent of adult T cell leukemia (ATL) and HTLV-1 associated myopathy/tropic spastic paraparesis (HAM/TSP). HTLV-1 expresses several regulatory and accessory genes that function at different stages of the virus life cycle. The regulatory gene Tax-1 is required for efficient virus replication, as it drives transcription of viral gene products, and has also been demonstrated to play a key role in the pathogenesis of the virus. Several studies have identified a PDZ binding motif (PBM) at the carboxyl terminus of Tax-1 and demonstrated the importance of this domain for HTLV-1 induced cellular transformation. Using a mass spectrometry-based proteomics approach we identified sorting nexin 27 (SNX27) as a novel interacting partner of Tax-1. Further, we demonstrated that their interaction is mediated by the Tax-1 PBM and SNX27 PDZ domains. SNX27 has been shown to promote the plasma membrane localization of glucose transport 1 (GLUT1), one of the receptor molecules of the HTLV-1 virus, and the receptor molecule required for HTLV-1 fusion and entry. We postulated that Tax-1 alters GLUT1 localization via its interaction with SNX27. We demonstrate that over expression of Tax-1 in cells causes a reduction of GLUT1 on the plasma membrane. Furthermore, we show that knockdown of SNX27 results in increased virion release and decreased HTLV-1 infectivity. Collectively, we demonstrate the first known mechanism by which HTLV-1 regulates a receptor molecule post-infection.',\n",
       "  '30897179'],\n",
       " 'Context-dependent scaling of kinematics and energetics during contests and feeding in mantis shrimp.': ['Measurements of energy use, and its scaling with size, are critical to understanding how organisms accomplish myriad tasks. For example, energy budgets are central to game theory models of assessment during contests and underlie patterns of feeding behavior. Clear tests connecting energy to behavioral theory require measurements of the energy use of single individuals for particular behaviors. Many species of mantis shrimp (Stomatopoda: Crustacea) use elastic energy storage to power high-speed strikes that they deliver to opponents during territorial contests and to hard-shelled prey while feeding. We compared the scaling of strike kinematics and energetics between feeding and contests in the mantis shrimp  We filmed strikes with high-speed video, measured strike velocity and used a mathematical model to calculate strike energy. During contests, strike velocity did not scale with body size but strike energy scaled positively with size. Conversely, while feeding, strike velocity decreased with increasing size and strike energy did not vary according to body size. Individuals most likely achieved this strike variation through differential compression of their exoskeletal spring prior to the strike.  analyses found that  used greater velocity and energy when striking larger opponents, yet variation in prey size was not accompanied by varying strike velocity or energetics. Our estimates of energetics inform prior tests of contest and feeding behavior in this species. More broadly, our findings elucidate the role behavioral context plays in measurements of animal performance.',\n",
       "  '30890620'],\n",
       " 'Dr. Hennis Green on Scrofulous Tubercle of the Brain in Children.': ['',\n",
       "  '21379958'],\n",
       " 'Dr. Hennis Green on Diseases of Children.': ['', '21372959'],\n",
       " 'Contributions to the Pathology of Children.': ['', '21379938'],\n",
       " 'On Hypertrophy of the Brain in Children.': ['', '21372866'],\n",
       " 'Intracranial arterial spasm in the dog. A chronic experimental model.': ['',\n",
       "  '5457638'],\n",
       " 'Potassium and the pathogenesis of cerebral arterial spasm in dog and man.': ['',\n",
       "  '5570771'],\n",
       " 'The effects of subarachnoid blood on the electrocorticogram of the cat.': ['',\n",
       "  '5570780'],\n",
       " 'Group selection on the boundary of a stable population.': ['A model of group selection is constructed for the case of differential extinction acting on small boundary populations of a large, fixed population. Consideration is restricted to extinction operators acting at or near to carrying capacity. Under the assumption that the extinction rate is large relative to individual genetic parameters affecting gene frequencies in boundary populations, we discuss the conditions under which differential extinction is most likely to produce a significant effect. In particular, a condition for bimodality in the distribution of gene frequencies in boundary populations (population polymorphism) is that there be some critical allele frequency at which the extinction rate jumps from high to low. An extinction operator linear in allele frequency produces no qualitative effect of this kind. In consequence, we are able to make precise rather limited circumstances under which group selection is likely to have a significant effect.',\n",
       "  '16592017'],\n",
       " 'A frequency-dependent natural selection model for the evolution of social cooperation networks.': ['A model is presented for the evolution of several aspects of sociality based on reciprocal ties of social cooperation, modeling especially cooperative hunting behavior in carnivores. This model captures the possibility of a critical threshold in gene frequency, which, if reached, will lead to an explosion toward fixation of the \"social\" trait. This threshold phenomenon might be restated as follows: the precondition for evolution favorable to the specific form of social behavior considered is hard to satisfy, but-once this condition is satisfied-the tendency toward sociality is effectively irreversible. The simple model proposed appears to be highly robust, with most realistic changes additionally favoring the social gene.',\n",
       "  '4509649'],\n",
       " 'Social Origins of Developmental Risk for Mental and Physical Illness.': [\"Adversity in early childhood exerts an enduring impact on mental and physical health, academic achievement, lifetime productivity, and the probability of interfacing with the criminal justice system. More science is needed to understand how the brain is affected by early life stress (ELS), which produces excessive activation of stress response systems broadly throughout the child's body (toxic stress). Our research examines the importance of sex, timing and type of stress exposure, and critical periods for intervention in various brain systems across species. Neglect (the absence of sensitive and responsive caregiving) or disrupted interaction with offspring induces robust, lasting consequences in mice, monkeys, and humans. Complementary assessment of internalizing disorders and brain imaging in children suggests that early adversity can interfere with white matter development in key brain regions, which may increase risk for emotional difficulties in the long term. Neural circuits that are most plastic during ELS exposure in monkeys sustain the greatest change in gene expression, offering a mechanism whereby stress timing might lead to markedly different long-term behaviors. Rodent models reveal that disrupted maternal-infant interactions yield metabolic and behavioral outcomes often differing by sex. Moreover, ELS may further accelerate or delay critical periods of development, which reflect GABA circuit maturation, BDNF, and circadian  genes. Such factors are associated with several mental disorders and may contribute to a premature closure of plastic windows for intervention following ELS. Together, complementary cross-species studies are elucidating principles of adaptation to adversity in early childhood with molecular, cellular, and whole organism resolution.\",\n",
       "  '29118206'],\n",
       " 'Developmental Connectivity and Molecular Phenotypes of Unique Cortical Projection Neurons that Express a Synapse-Associated Receptor Tyrosine Kinase.': ['The complex circuitry and cell-type diversity of the cerebral cortex are required for its high-level functions. The mechanisms underlying the diversification of cortical neurons during prenatal development have received substantial attention, but understanding of neuronal heterogeneity is more limited during later periods of cortical circuit maturation. To address this knowledge gap, connectivity analysis and molecular phenotyping of cortical neuron subtypes that express the developing synapse-enriched MET receptor tyrosine kinase were performed. Experiments used a MetGFP transgenic mouse line, combined with coexpression analysis of class-specific molecular markers and retrograde connectivity mapping. The results reveal that MET is expressed by a minor subset of subcerebral and a larger number of intratelencephalic projection neurons. Remarkably, MET is excluded from most layer 6 corticothalamic neurons. These findings are particularly relevant for understanding the maturation of discrete cortical circuits, given converging evidence that MET influences dendritic elaboration and glutamatergic synapse maturation. The data suggest that classically defined cortical projection classes can be further subdivided based on molecular characteristics that likely influence synaptic maturation and circuit wiring. Additionally, given that MET is classified as a high confidence autism risk gene, the data suggest that projection neuron subpopulations may be differentially vulnerable to disorder-associated genetic variation.',\n",
       "  '29190358'],\n",
       " 'Differential impact of <i>Met</i> receptor gene interaction with early-life stress on neuronal morphology and behavior in mice.': [\"Early adversity in childhood increases the risk of anxiety, mood, and post-traumatic stress disorders in adulthood, and specific gene-by-environment interactions may increase risk further. A common functional variant in the promoter region of the gene encoding the human MET receptor tyrosine kinase (rs1858830 ' allele) reduces expression of  and is associated with altered cortical circuit function and structural connectivity. Mice with reduced  expression exhibit changes in anxiety-like and conditioned fear behavior, precocious synaptic maturation in the hippocampus, and reduced neuronal arbor complexity and synaptogenesis. These phenotypes also can be produced independently by early adversity in wild-type mice. The present study addresses the outcome of combining early-life stress and genetic influences that alter timing of maturation on enduring functional and structural phenotypes. Using a model of reduced  expression ( ) and early-life stress from postnatal day 2-9, social, anxiety-like, and contextual fear behaviors in later life were measured. Mice that experienced early-life stress exhibited impairments in social interaction, whereas alterations in anxiety-like behavior and fear learning were driven by  haploinsufficiency, independent of rearing condition. Early-life stress or reduced  expression decreased arbor complexity of ventral hippocampal CA1 pyramidal neurons projecting to basolateral amygdala. Paradoxically, arbor complexity in  mice was increased following early-life stress, and thus not different from arbors in wild-type mice raised in control conditions. The changes in dendritic morphology are consistent with the hypothesis that the physiological state of maturation of CA1 neurons in  mice influences their responsiveness to early-life stress. The dissociation of behavioral and structural changes suggests that there may be phenotype-specific sensitivities to early-life stress.\",\n",
       "  '29255778'],\n",
       " 'Analysis of shared heritability in common disorders of the brain.': ['Disorders of the brain can exhibit considerable epidemiological comorbidity and often share symptoms, provoking debate about their etiologic overlap. We quantified the genetic sharing of 25 brain disorders from genome-wide association studies of 265,218 patients and 784,643 control participants and assessed their relationship to 17 phenotypes from 1,191,588 individuals. Psychiatric disorders share common variant risk, whereas neurological disorders appear more distinct from one another and from the psychiatric disorders. We also identified significant sharing between disorders and a number of brain phenotypes, including cognitive measures. Further, we conducted simulations to explore how statistical power, diagnostic misclassification, and phenotypic heterogeneity affect genetic correlations. These results highlight the importance of common genetic variation as a risk factor for brain disorders and the value of heritability-based methods in understanding their etiology.',\n",
       "  '29930110'],\n",
       " 'Integrative functional genomic analysis of human brain development and neuropsychiatric risks.': ['To broaden our understanding of human neurodevelopment, we profiled transcriptomic and epigenomic landscapes across brain regions and/or cell types for the entire span of prenatal and postnatal development. Integrative analysis revealed temporal, regional, sex, and cell type-specific dynamics. We observed a global transcriptomic cup-shaped pattern, characterized by a late fetal transition associated with sharply decreased regional differences and changes in cellular composition and maturation, followed by a reversal in childhood-adolescence, and accompanied by epigenomic reorganizations. Analysis of gene coexpression modules revealed relationships with epigenomic regulation and neurodevelopmental processes. Genes with genetic associations to brain-based traits and neuropsychiatric disorders (including , , , , and ) converged in a small number of modules and distinct cell types, revealing insights into neurodevelopment and the genomic basis of neuropsychiatric risks.',\n",
       "  '30545854'],\n",
       " 'Resolving the full spectrum of human genome variation using Linked-Reads.': ['Large-scale population analyses coupled with advances in technology have demonstrated that the human genome is more diverse than originally thought. To date, this diversity has largely been uncovered using short-read whole-genome sequencing. However, these short-read approaches fail to give a complete picture of a genome. They struggle to identify structural events, cannot access repetitive regions, and fail to resolve the human genome into haplotypes. Here, we describe an approach that retains long range information while maintaining the advantages of short reads. Starting from ∼1 ng of high molecular weight DNA, we produce barcoded short-read libraries. Novel informatic approaches allow for the barcoded short reads to be associated with their original long molecules producing a novel data type known as \"Linked-Reads\". This approach allows for simultaneous detection of small and large variants from a single library. In this manuscript, we show the advantages of Linked-Reads over standard short-read approaches for reference-based analysis. Linked-Reads allow mapping to 38 Mb of sequence not accessible to short reads, adding sequence in 423 difficult-to-sequence genes including disease-relevant genes , , and  Both Linked-Read whole-genome and whole-exome sequencing identify complex structural variations, including balanced events and single exon deletions and duplications. Further, Linked-Reads extend the region of high-confidence calls by 68.9 Mb. The data presented here show that Linked-Reads provide a scalable approach for comprehensive genome analysis that is not possible using short reads alone.',\n",
       "  '30894395'],\n",
       " 'Genomes for all.': ['', '30188541'],\n",
       " 'Corrigendum: Direct determination of diploid genome sequences.': ['',\n",
       "  '29610250'],\n",
       " 'Dissecting the Causal Mechanism of X-Linked Dystonia-Parkinsonism by Integrating Genome and Transcriptome Assembly.': ['X-linked Dystonia-Parkinsonism (XDP) is a Mendelian\\xa0neurodegenerative disease that is endemic to the Philippines and is associated with a founder haplotype. We integrated multiple genome and transcriptome assembly technologies to narrow the causal mutation to the TAF1 locus, which included a SINE-VNTR-Alu (SVA) retrotransposition into intron 32 of the gene. Transcriptome analyses identified decreased expression of the canonical cTAF1 transcript among XDP probands, and de novo assembly\\xa0across multiple pluripotent stem-cell-derived neuronal lineages discovered aberrant TAF1 transcription that involved alternative splicing and intron retention (IR) in proximity to the SVA that was anti-correlated with overall TAF1 expression. CRISPR/Cas9 excision of the SVA rescued this XDP-specific\\xa0transcriptional signature and normalized TAF1 expression in probands. These data suggest an SVA-mediated aberrant transcriptional mechanism associated with XDP and may provide a roadmap for layered technologies and integrated assembly-based analyses for other unsolved Mendelian disorders.',\n",
       "  '29474918'],\n",
       " 'Reference quality assembly of the 3.5-Gb genome of <i>Capsicum annuum</i> from a single linked-read library.': ['Linked-Read sequencing technology has recently been employed successfully for  assembly of human genomes, however, the utility of this technology for complex plant genomes is unproven. We evaluated the technology for this purpose by sequencing the 3.5-gigabase (Gb) diploid pepper () genome with a single Linked-Read library. Plant genomes, including pepper, are characterized by long, highly similar repetitive sequences. Accordingly, significant effort is used to ensure that the sequenced plant is highly homozygous and the resulting assembly is a haploid consensus. With a phased assembly approach, we targeted a heterozygous F derived from a wide cross to assess the ability to derive both haplotypes and characterize a pungency gene with a large insertion/deletion. The Supernova software generated a highly ordered, more contiguous sequence assembly than all currently available  reference genomes. Over 83% of the final assembly was anchored and oriented using four publicly available\\xa0 linkage maps. A comparison of the annotation of conserved eukaryotic genes indicated the completeness of assembly. The validity of the phased assembly is further demonstrated with the complete recovery of both 2.5-Kb insertion/deletion haplotypes of the  locus in the F sample that represents pungent and nonpungent peppers, as well as nearly full recovery of the BUSCO2 gene set within each of the two haplotypes. The most contiguous pepper genome assembly to date has been generated which demonstrates that Linked-Read library technology provides a tool to  assemble complex highly repetitive heterozygous plant genomes. This technology can provide an opportunity to cost-effectively develop high-quality genome assemblies for other complex plants and compare structural and gene differences through accurate haplotype reconstruction.',\n",
       "  '29423234'],\n",
       " 'The Ras GTPase-activating-protein-related human protein IQGAP2 harbors a potential actin binding domain and interacts with calmodulin and Rho family GTPases.': ['We previously described IQGAP1 as a human protein related to a putative Ras GTPase-activating protein (RasGAP) from the fission yeast Schizosaccharomyces pombe. Here we report the identification of a liver-specific human protein that is 62% identical to IQGAP1. Like IQGAP1, the novel IQGAP2 protein harbors an N-terminal calponin homology motif which functions as an F-actin binding domain in members of the spectrin, filamin, and fimbrin families. Both IQGAPs also harbor several copies of a novel 50- to 55-amino-acid repeat, a single WW domain, and four IQ motifs and have 25% sequence identity with almost the entire S. pombe sar1 RasGAP homolog. As predicted by the presence of IQ motifs, IQGAP2 binds calmodulin. However, neither full-length nor truncated IQGAP2 stimulated the GTPase activity of Ras or its close relatives. Instead, IQGAP2 binds Cdc42 and Racl but not RhoA. This interaction involves the C-terminal half of IQGAP2 and appears to be independent of the nucleotide binding status of the GTPases. Although IQGAP2 shows no GAP activity towards Cdc42 and Rac1, the protein did inhibit both the intrinsic and RhoGAP-stimulated GTP hydrolysis rates of Cdc42 and Rac1, suggesting an alternative mechanism via which IQGAPs might modulate signaling by these GTPases. Since IQGAPs harbor a potential actin binding domain, they could play roles in the Cdc42 and Rac1 controlled generation of specific actin structures.',\n",
       "  '8756646'],\n",
       " 'Molecular definition of deletions of different segments of distal 5p that result in distinct phenotypic features.': ['Cri du chat syndrome (CDC) is a segmental aneusomy associated with deletions of chromosome 5p15. In an effort to define regions that produce the phenotypes associated with CDC, we have analyzed deletions from 17 patients. The majority of these patients had atypical CDC features or were asymptomatic. Using these patients, we have mapped several phenotypes associated with deletions of 5p, including speech delay, catlike cry, newborn facial dysmorphism, and adult facial dysmorphism. This phenotypic map should provide a framework with which to begin identification of genes associated with various phenotypic features associated with deletions of distal 5p. We have also analyzed the parental origin of the de novo deletions, to determine if genomic imprinting could be occurring in this region. In addition, we have isolated cosmids that could be useful for both prenatal and postnatal assessments of del5(p) individuals.',\n",
       "  '7726173'],\n",
       " 'Mutations in the transmembrane domain of FGFR3 cause the most common genetic form of dwarfism, achondroplasia.': [\"Achondroplasia (ACH) is the most common genetic form of dwarfism. This disorder is inherited as an autosomal dominant trait, although the majority of cases are sporadic. A gene for ACH was recently localized to 4p16.3 by linkage analyses. The ACH candidate region includes the gene encoding fibroblast growth factor receptor 3 (FGFR3), which was originally considered as a candidate for the Huntington's disease gene. DNA studies revealed point mutations in the FGFR3 gene in ACH heterozygotes and homozygotes. The mutation on 15 of the 16 ACH-affected chromosomes was the same, a G-->A transition, at nucleotide 1138 of the cDNA. The mutation on the only ACH-affected chromosome 4 without the G-->A transition at nucleotide 1138 had a G-->C transversion at this same position. Both mutations result in the substitution of an arginine residue for a glycine at position 380 of the mature protein, which is in the transmembrane domain of FGFR3.\",\n",
       "  '7913883'],\n",
       " 'Isolation of genes from complex sources of mammalian genomic DNA using exon amplification.': ['Modifications to exon amplification have been instituted that increase its speed, efficiency and reliability. Exons were isolated from target human or mouse genomic DNA sources ranging from 30 kilobases (kb) to 3 megabases (Mb) in complexity. The efficiency was dependent upon the amount of input DNA, and ranged from isolation of an exon for every 20 kb to an exon for every 80 kb of target genomic DNA. In these studies, several novel genes and a smaller number of genes isolated previously that reside on human chromosome 9 have been identified. These results indicate that exon amplification is presently adaptable to large scale isolation of exons from complex sources of genomic DNA.',\n",
       "  '8136842'],\n",
       " 'Identification of human chromosome 9 specific genes using exon amplification.': ['We have recently developed a method, exon amplification, that is designed for isolation of exon sequences from genomic DNA. To assess the efficacy of this method we have analyzed cosmid genomic clones derived from human chromosome 9, and have cloned several products from this analysis. Approximately 63% of cosmids produced at least one product derived from functioning splice sites within the target genomic fragment, and in many cases multiple products were isolated. In addition, an easily identifiable class of false positives was produced from 56% of cosmids analyzed; these are readily eliminated from subsequent study. Sequence analysis and database searches revealed that the majority (87%) of the putative exon clones were unique, the remainder being derived from repetitive sequences. Analysis of sequence conservation by Southern blotting in addition to cDNA screening experiments suggested that most, if not all, of these unique sequences represent true exons. The results of these studies indicate that exon amplification is a rapid and reliable approach for isolation of exon sequences from mammalian genomic DNA.',\n",
       "  '7506603'],\n",
       " 'Differential alveolar bone modeling after orthodontic retraction.': ['BACKGROUND: Distinct, irregular, and hard nodular protuberances similar to the morphologic features of exostoses can occasionally be noted on the labial surface of the alveolar bone after orthodontic retraction of anterior teeth in adults. These have long been believed to be exostoses developed in response to loading. However, specific characterization of this phenomenon has not been documented.\\n\\nCASE DESCRIPTIONS: Three cases of patients displaying multiple irregular labial bony protuberances after retraction of anterior teeth are reported. These protuberances appeared during retraction and became more prominent with additional retraction. Serial clinical photographs, lateral cephalograms, digital models, and cone-beam computed tomography scans were evaluated. On the basis of 3-dimensional superimpositions of digital models and cone-beam computed tomographic scans, the irregular protuberances appear to be the result of differential alveolar bone modeling, with more resorption of bone covering the tooth root than that of interdental bone, and not of true bone overgrowth or deposition (that is, exostoses).\\n\\nCONCLUSION AND PRACTICAL IMPLICATIONS: Orthodontic patients often seek treatment to improve occlusion as well as esthetics. Although this study shows that these protuberances are the result of differential modeling, they may still be perceived by patients as \"outgrowths,\" which may cause concerns related to esthetics or comfort. Clinicians should note that these protuberances are a possible outcome when large amounts of bodily retraction and root movement of anterior teeth are planned. Patients who experience psychosocial problems with this phenomenon may be candidates for alveoloplasty.',\n",
       "  '30922461'],\n",
       " 'Intraoperative hypotension is a risk factor for postoperative acute kidney injury after femoral neck fracture surgery: a retrospective study.': ['BACKGROUND: Hip fracture in elderly patients is a serious health concern due to the associated morbidity and mortality. Although acute kidney injury after hip fracture is known to be a significantly poor prognostic factor for morbidity and mortality, the literature regarding the risk factors for acute kidney injury after hip fracture is insufficient. This study aimed to investigate the incidence and associated risk factors for acute kidney injury in patients with femoral neck fracture.\\n\\nMETHODS: A total of 248 patients who underwent an operation for femoral neck fracture between January 2011 and January 2015 were retrospectively analyzed. Acute kidney injury was defined according to the Kidney Disease: Improving Global Outcomes guidelines.\\n\\nRESULTS: The incidence of acute kidney injury was 17.7% (n\\u2009=\\u200944). Risk factors for acute kidney injury included diabetes mellitus, pre-existing renal disease, preoperative blood urea nitrogen (BUN), preoperative estimated glomerular filtration rate (eGFR), preoperative haemoglobin (Hb) level, type of operation, postoperative creatinine level and intraoperative hypotension (P\\u2009<\\u2009\\xa00.05). After controlling for confounding variables, intraoperative hypotension was only the independent risk factor for acute kidney injury (P\\u2009=\\u20090.012).\\n\\nCONCLUSIONS: Acute kidney injury was found to occur frequently after surgery for femur neck fracture. Surgeons should be aware of acute kidney injury when planning the management of patients with femoral neck fracture and consider that the duration of intraoperative hypotension is a risk factor for acute kidney injury.',\n",
       "  '30917804'],\n",
       " 'Hereditary spherocytosis caused by copy number variation in SPTB gene identified through targeted next-generation sequencing.': [\"Hereditary spherocytosis (HS) is a heterogeneous genetic disorder characterized by spherocytosis on peripheral blood smear with hemolytic anemia, accompanied by signs of hemolysis. Herein, we report a 5-month-old Korean girl with HS resulting from a de novo 271\\xa0Kb microdeletion of 14q23.3. She presented with hemolytic anemia and mild splenomegaly. Spherocytosis was seen on examination of peripheral blood. Eosin-5'-maleimide (EMA) test and flow cytometric osmotic fragility test were positive. She had no relevant family history of spherocytosis. No pathogenic single nucleotide variants or small insertions/deletions were detected in HS-associated genes. Array comparative genomic hybridization analysis revealed a 271\\xa0Kb deletion at chromosome 14q23.3, encompassing the SPTB, CHURC1, GPX2, RAB15, FNTB, and MAX genes. We found a deletion affecting 5' UTR, exon 1, and part of intron 1 of the SPTB gene using targeted next-generation sequencing (NGS) analysis, suggesting that NGS may be able to identify disease-causing copy number variations (CNVs), as well as small point mutations in HS patients. In addition, chromosomal microarray may be useful in defining combined deleted genes. Additional evaluations should thus be considered in the diagnosis of HS, especially when CNV is revealed as disease-causing abnormality.\",\n",
       "  '30903564']}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dic = {}\n",
    "for pub in abs_corpus: \n",
    "    dic[pub.title] = [pub.abstract,pub.pmid]\n",
    "dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('more_corp.csv', 'w') as f:  # Just use 'w' mode in 3.x\n",
    "    w = csv.DictWriter(f, dic.keys())\n",
    "    w.writeheader()\n",
    "    w.writerow(dic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other sources:\n",
    "\n",
    "Other popular sources for internet data:\n",
    "\n",
    "[reddit](https://www.reddit.com/) - https://praw.readthedocs.io/en/v2.1.21/\n",
    "\n",
    "[twitter](https://twitter.com/) - https://pypi.org/project/python-twitter/\n",
    "\n",
    "[project gutenburg](https://www.gutenberg.org/) - https://github.com/ageitgey/Gutenberg \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
